{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ee2c941b-c657-47eb-9ff0-b0ab16fe7a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# dir\n",
    "# DATASET_DIR = Path('/mnt/f/spatial_data/probe')\n",
    "DATASET_DIR = Path('./probe')\n",
    "RUNID = 'example_dataset'\n",
    "workdir = DATASET_DIR / RUNID\n",
    "os.makedirs(workdir, exist_ok=True)\n",
    "organism = 'mouse'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "74fbaff4-839a-4a74-87ac-74ed2904805e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create results dir\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "output = os.path.join(workdir, 'results', formatted_time+'_ensembl')\n",
    "bds_candidate_dir = os.path.join(output, \"bds_candidate\")\n",
    "os.makedirs(output, exist_ok=True)\n",
    "os.makedirs(bds_candidate_dir, exist_ok=True)\n",
    "\n",
    "# file name variables\n",
    "bds_candidate_file_suffix = \"_bds_candidate.fasta\"\n",
    "combined_bds_candidates_file = \"total_bds_candidate.fasta\"\n",
    "combined_bds_candidates_blast_file = 'total_bds_candidate_blast.fasta'\n",
    "bds_candidate_num_file = \"bds_candidate_num.json\"\n",
    "blast_results_file = \"blast_results.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "418225f4-2503-4339-a36f-4cd33cbebd35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gapdh</td>\n",
       "      <td>Gapdh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Actb</td>\n",
       "      <td>Actb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hprt1</td>\n",
       "      <td>Hprt1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name   gene\n",
       "0     Gapdh  Gapdh\n",
       "1      Actb   Actb\n",
       "2     Hprt1  Hprt1"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene_info = pd.read_excel(os.path.join(workdir, \"gene_info.xlsx\"))\n",
    "gene_info = pd.read_table(os.path.join(workdir, \"gene_list_JC.txt\"))\n",
    "gene_info['gene'] = gene_info['gene_name']\n",
    "if organism == 'mouse': gene_info['gene'] = gene_info['gene'].str.capitalize()\n",
    "elif organism == 'human': gene_info['gene'] = gene_info['gene'].str.upper()\n",
    "gene_list = [_.strip() for _ in gene_info['gene'].unique() if _!=0]\n",
    "print(len(gene_list))\n",
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d9a08a65-0c1a-486f-9a76-9f5216ac306d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retriving_sequences:   0%|          | 0/3 [00:00<?, ?it/s]\n",
      "Gapdh:   0%|          | 0/20 [00:00<?, ?it/s]\u001b[A\n",
      "Gapdh:   5%|▌         | 1/20 [00:00<00:07,  2.58it/s]\u001b[A\n",
      "Gapdh:  10%|█         | 2/20 [00:00<00:06,  2.73it/s]\u001b[A\n",
      "Gapdh:  15%|█▌        | 3/20 [00:01<00:06,  2.67it/s]\u001b[A\n",
      "Gapdh:  20%|██        | 4/20 [00:01<00:06,  2.64it/s]\u001b[A\n",
      "Gapdh:  25%|██▌       | 5/20 [00:01<00:05,  2.57it/s]\u001b[A\n",
      "Gapdh:  30%|███       | 6/20 [00:03<00:08,  1.59it/s]\u001b[A\n",
      "Gapdh:  35%|███▌      | 7/20 [00:03<00:07,  1.76it/s]\u001b[A\n",
      "Gapdh:  40%|████      | 8/20 [00:03<00:06,  1.99it/s]\u001b[A\n",
      "Gapdh:  45%|████▌     | 9/20 [00:04<00:05,  2.08it/s]\u001b[A\n",
      "Gapdh:  50%|█████     | 10/20 [00:04<00:04,  2.03it/s]\u001b[A\n",
      "Gapdh:  55%|█████▌    | 11/20 [00:05<00:04,  2.02it/s]\u001b[A\n",
      "Gapdh:  60%|██████    | 12/20 [00:05<00:04,  1.99it/s]\u001b[A\n",
      "Gapdh:  65%|██████▌   | 13/20 [00:06<00:03,  2.04it/s]\u001b[A\n",
      "Gapdh:  70%|███████   | 14/20 [00:06<00:02,  2.20it/s]\u001b[A\n",
      "Gapdh:  75%|███████▌  | 15/20 [00:07<00:02,  2.31it/s]\u001b[A\n",
      "Gapdh:  80%|████████  | 16/20 [00:07<00:01,  2.35it/s]\u001b[A\n",
      "Gapdh:  85%|████████▌ | 17/20 [00:07<00:01,  2.39it/s]\u001b[A\n",
      "Gapdh:  90%|█████████ | 18/20 [00:08<00:00,  2.43it/s]\u001b[A\n",
      "Gapdh:  95%|█████████▌| 19/20 [00:08<00:00,  2.48it/s]\u001b[A\n",
      "Gapdh: 100%|██████████| 20/20 [00:09<00:00,  2.38it/s]\u001b[A\n",
      "Retriving_sequences:  33%|███▎      | 1/3 [00:10<00:20, 10.05s/it]\n",
      "Actb:   0%|          | 0/9 [00:00<?, ?it/s]\u001b[A\n",
      "Actb:  11%|█         | 1/9 [00:00<00:03,  2.11it/s]\u001b[A\n",
      "Actb:  22%|██▏       | 2/9 [00:00<00:03,  2.31it/s]\u001b[A\n",
      "Actb:  33%|███▎      | 3/9 [00:01<00:02,  2.39it/s]\u001b[A\n",
      "Actb:  44%|████▍     | 4/9 [00:01<00:02,  2.41it/s]\u001b[A\n",
      "Actb:  56%|█████▌    | 5/9 [00:02<00:01,  2.28it/s]\u001b[A\n",
      "Actb:  67%|██████▋   | 6/9 [00:02<00:01,  1.97it/s]\u001b[A\n",
      "Actb:  78%|███████▊  | 7/9 [00:03<00:00,  2.01it/s]\u001b[A\n",
      "Actb:  89%|████████▉ | 8/9 [00:03<00:00,  2.04it/s]\u001b[A\n",
      "Actb: 100%|██████████| 9/9 [00:04<00:00,  2.15it/s]\u001b[A\n",
      "Retriving_sequences:  67%|██████▋   | 2/3 [00:15<00:07,  7.22s/it]\n",
      "Hprt1:   0%|          | 0/1 [00:00<?, ?it/s]\u001b[A\n",
      "Hprt1: 100%|██████████| 1/1 [00:00<00:00,  2.34it/s]\u001b[A\n",
      "Retriving_sequences: 100%|██████████| 3/3 [00:16<00:00,  5.55s/it]\n"
     ]
    }
   ],
   "source": [
    "from lib.database_interaction import ensembl_name_to_seqs\n",
    "import time\n",
    "max_trial = 3\n",
    "sequences_of_all = dict()\n",
    "error_messages = {gene: [] for gene in gene_list}\n",
    "\n",
    "with tqdm(total=len(gene_list), desc=\"Retriving_sequences\", position=0) as pbar_total:\n",
    "    for gene in gene_list:\n",
    "        sequences_of_all[gene] = {}\n",
    "        trial_success = False\n",
    "        \n",
    "        # Reset the trial progress bar for each gene\n",
    "        for trial in range(1, max_trial+1):  # Retrying up to 3 times\n",
    "            try:\n",
    "                # Attempt to retrieve sequences\n",
    "                sequences_of_all[gene] = ensembl_name_to_seqs(gene=gene, species=organism, seq_type='cds', tqdm_args={'position': 1,'leave': False})\n",
    "                trial_success = True\n",
    "                break\n",
    "            except Exception as e:\n",
    "                time.sleep(1)\n",
    "\n",
    "        if not trial_success:\n",
    "            error_messages[gene].append(f\"Failed to retrieve sequences for {gene} after {max_trial} attempts.\")\n",
    "\n",
    "        pbar_total.update(1)  # Update the main progress bar after each gene\n",
    "\n",
    "for gene, messages in error_messages.items():\n",
    "    for message in messages:\n",
    "        print(message)\n",
    "\n",
    "with open(os.path.join(output, 'sequence_of_all.json'), 'w') as file: json.dump(sequences_of_all, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2c4c4b64-5cde-48ec-9c4b-94b91c12ec5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(output, 'sequence_of_all.json'), 'r') as file: sequences_of_all = json.load(file)\n",
    "\n",
    "isoforms = {}\n",
    "for gene, isoforms_tmp in sequences_of_all.items():\n",
    "    shortest_isoform = None\n",
    "    min_length = 10**6\n",
    "    max_length = 0\n",
    "    for isoform in isoforms_tmp:\n",
    "        try: isoform_name = isoform['external_name']\n",
    "        except: continue\n",
    "        if len(isoform['seq']) < min_length and gene.upper() in isoform_name.upper():\n",
    "            min_length = len(isoform['seq'])\n",
    "            shortest_isoform = isoform\n",
    "    if shortest_isoform:\n",
    "        isoforms[gene] = shortest_isoform\n",
    "with open(os.path.join(output, 'shortest_isoforms.json'), 'w') as file: json.dump(isoforms, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0de0affe-bae4-46c6-bd5a-fa3b773ecad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 查找绑定位点"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c91e620a-13f2-442f-8587-0399edf90168",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Searching_binding_sites:   0%|          | 0/3 [00:00<?, ?it/s]\n",
      "Gapdh-207:   0%|          | 0/109 [00:00<?, ?it/s]\u001b[A\n",
      "                                                  \u001b[A/tmp/ipykernel_1617292/3508141571.py:48: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  BDS_INFO = pd.concat([BDS_INFO, add], ignore_index=True)\n",
      "Searching_binding_sites:  33%|███▎      | 1/3 [00:00<00:00,  3.06it/s]\n",
      "Actb-203:   0%|          | 0/252 [00:00<?, ?it/s]\u001b[A\n",
      "                                                 \u001b[A\n",
      "Hprt1-201:   0%|          | 0/553 [00:00<?, ?it/s]\u001b[A\n",
      "Searching_binding_sites: 100%|██████████| 3/3 [00:00<00:00,  6.83it/s]\n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# set lib auto reload in jupyter notebook\n",
    "from lib.search_binding import position_search, optimize_subsequence, seq_minus\n",
    "\n",
    "# Initiation of array\n",
    "binding_site_entry = [\n",
    "    \"accession\", \"gene_name\", \"mol_type\", \"organism\",\n",
    "    \"pos\", \"plp_bds\", \"plp_Tm\",\"plp_bds3'\", \"plp_bds5'\", \"plp_Tm3'\", \"plp_Tm5'\", \"mfe\", \"wanted\"] ## 构建表头\n",
    "alignment_entry = [\"align_num\", \"align_accession\", \"align_descrip\", \"plus/minus\"]\n",
    "BDS_INFO = pd.DataFrame(columns=binding_site_entry+alignment_entry)\n",
    "\n",
    "# Search binding sites on mRNA sequence\n",
    "pre_binding_num = {}\n",
    "\n",
    "# initialization of file\n",
    "with open(os.path.join(output, combined_bds_candidates_file), \"w\") as handle: handle.write(\"\")\n",
    "with open(os.path.join(output, combined_bds_candidates_blast_file), \"w\") as f: f.write(\"\")\n",
    "\n",
    "for desc, info in tqdm(isoforms.items(), desc=\"Searching_binding_sites\", position=0):\n",
    "    seq = info['seq']\n",
    "    if 'N' in seq: seq = seq.replace('N', '')\n",
    "    try: gene_name= info['external_name']\n",
    "    except: gene_name = desc\n",
    "\n",
    "    id = info['id']\n",
    "    mol_type = info['biotype']\n",
    "    \n",
    "    pos_info = position_search(\n",
    "        seq, gene=gene_name,\n",
    "        BDS_len=40, BDS_num=100, min_gap=0, better_gap=40, pin_gap=0.05, \n",
    "        G_min=0.25, G_max=0.7, G_consecutive=5, Tm_low=48, Tm_high=60, \n",
    "        verbose_pos=1, leave=False, warn=False) #设置筛选条件\n",
    "    \n",
    "    record_list = []\n",
    "    for i, pre_binding_tmp in enumerate([_['plp_bds'] for _ in pos_info]):\n",
    "        record_list.append(\n",
    "            SeqRecord(Seq(pre_binding_tmp), id=\"bds_candidate\" + str(i), \n",
    "                      description=\"|\".join([id, gene_name, organism, mol_type])))\n",
    "\n",
    "    # add information about binding sites to FOI\n",
    "    add = pd.DataFrame(pos_info)\n",
    "    add['accession'] = id\n",
    "    add['gene_name'] = gene_name\n",
    "    add['mol_type'] = mol_type\n",
    "    add['organism'] = organism\n",
    "    BDS_INFO = pd.concat([BDS_INFO, add], ignore_index=True)\n",
    "\n",
    "    file_out = os.path.join(bds_candidate_dir, gene_name + bds_candidate_file_suffix)\n",
    "    \n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list: SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(os.path.join(output, combined_bds_candidates_file), \"a\") as handle:\n",
    "        for new_record in record_list: SeqIO.write(new_record, handle, \"fasta\")\n",
    "    with open(os.path.join(output, combined_bds_candidates_blast_file), \"a\") as handle:\n",
    "        for new_record in record_list: \n",
    "            blast_seq = str(new_record.seq)\n",
    "            blast_seq = blast_seq[len(blast_seq)//2-16:len(blast_seq)//2+16] #提取中心32 nt生成BLAST专用文件\n",
    "            new_record = SeqRecord(Seq(blast_seq), id=new_record.id, description=new_record.description)\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[f\"{id}_{gene_name}\"] = len(pos_info)\n",
    "\n",
    "with open(os.path.join(output, bds_candidate_num_file), \"w\") as f: json.dump(pre_binding_num, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "613b4f9b-bc03-4ef8-840f-5656646523c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "##需要导出fasta文件，用blast网站处理：\n",
    "##https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&BLAST_SPEC=&LINK_LOC=blasttab&LAST_PAGE=tblastn\n",
    "##输出后修改文件名为blast_results.xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8898fcbb-15b9-46f9-b95a-d19d37cd16e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract interested information from blast_results 提取blast的结果\n",
    "from Bio.Blast import NCBIXML\n",
    "align_num = []\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(os.path.join(output, blast_results_file), \"r\") as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split(\"|\")\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        BDS_INFO.loc[loca, \"align_accession\"] = \"|\".join(str(_) for _ in align_accession)\n",
    "        # add align_descrip to df\n",
    "        BDS_INFO.loc[loca, \"align_descrip\"] = \"|\".join(str(_) for _ in align_descrip_list)\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "        # add plus/minus to df\n",
    "        try: BDS_INFO.loc[loca, \"plus/minus\"] = \",\".join([str(_) for _ in p_m])\n",
    "        except: BDS_INFO.loc[loca, \"plus/minus\"] = pd.NA\n",
    "        loca += 1\n",
    "BDS_INFO[\"align_num\"] = align_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "38da389e-2ae9-429f-913e-48244dddce45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Gapdh-207: Not enough pos for 8 binding sites.\n",
      "Gene Gapdh-207: condition too harsh, loose to get better results\n",
      "[27, 28, 31, 38, 39, 46, 47, 51, 55, 75]\n",
      "Gene Actb-203: Not enough pos for 8 binding sites.\n",
      "Gene Actb-203: condition too harsh, loose to get better results\n",
      "[59, 60, 112, 120, 124, 125, 126, 127, 128, 129, 132, 236, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 254, 257, 258, 259]\n",
      "Gene Hprt1-201: Not enough pos for 8 binding sites.\n",
      "Gene Hprt1-201: condition too harsh, loose to get better results\n",
      "[112, 113, 114, 115, 127, 129, 130, 131, 132, 133, 134, 146, 147, 148, 149, 296, 447, 450, 453, 454, 457, 458, 459, 460, 470, 473, 474, 475, 476, 477, 478, 479, 480, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 499, 500]\n"
     ]
    }
   ],
   "source": [
    "## 选择结合位点\n",
    "import re\n",
    "\n",
    "def adjust_gene_name(gene_name, gene_list):\n",
    "    gene_list = [x.upper() for x in gene_list]\n",
    "    match = re.search(r'(.+)-(\\d+)$', gene_name)\n",
    "    if match:\n",
    "        base_gene_name = match.group(1)\n",
    "        if base_gene_name.upper() in gene_list or gene_name.upper() in gene_list: return base_gene_name\n",
    "        else: return gene_name\n",
    "    else: return gene_name\n",
    "## 原文输出probes_wanted.xlsx文档 因为.xlsx的文档在jupyter lab中很难打开\n",
    "## 于是尝试修改输出csv文件,成功\n",
    "BDS_INFO[\"wanted\"] = [True] * len(BDS_INFO)\n",
    "verbose = True\n",
    "# select by specifity\n",
    "gene_name_list = [_.upper() for _ in gene_list]\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(BDS_INFO)):\n",
    "    # check gene_name\n",
    "    gene_name = adjust_gene_name(BDS_INFO.loc[i, \"gene_name\"], gene_name_list)\n",
    "    spe_ori= BDS_INFO.loc[i, \"organism\"]\n",
    "    if gene_name.upper() not in gene_name_list: \n",
    "        BDS_INFO.loc[i, \"wanted\"] = False\n",
    "        if verbose: print(f\"{gene_name} not in gene list.\")\n",
    "    else:\n",
    "        try: gene_name_list_out.remove(gene_name)\n",
    "        except: pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        if BDS_INFO.loc[i, \"mol_type\"] != \"protein_coding\":\n",
    "            # BDS_INFO.loc[i, \"wanted\"] = False\n",
    "            if verbose: print(\"{} is {}.\".format(gene_name, BDS_INFO.loc[i, \"mol_type\"]))\n",
    "\n",
    "    # check gene_organism name\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        descrip = BDS_INFO.loc[i, \"align_descrip\"]\n",
    "        if pd.isnull(descrip):\n",
    "            BDS_INFO.loc[i, \"wanted\"] = False\n",
    "            if verbose: print(f\"{gene_name} not found in BLAST.\")\n",
    "        else:\n",
    "            descrip = descrip.split(\"|\")\n",
    "            for des in descrip:\n",
    "                if gene_name not in des and spe_ori in des:\n",
    "                    BDS_INFO.loc[i, \"wanted\"] = False\n",
    "                    if verbose: print(f\"{gene_name} not specific.\")\n",
    "                    break\n",
    "\n",
    "    # check plus/minus\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        pm_list = BDS_INFO.loc[i, \"plus/minus\"].split(\",\")\n",
    "        if \"-1\" not in pm_list:\n",
    "            BDS_INFO.loc[i, \"wanted\"] = False\n",
    "            if verbose: print(f\"{gene_name} not plus/minus.\")\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "BDS_INFO.to_csv(os.path.join(output, \"probes_candidates.csv\"))\n",
    "\n",
    "out_tmp = BDS_INFO[BDS_INFO[\"wanted\"] == True]\n",
    "output_df = pd.DataFrame()\n",
    "for gene in out_tmp.gene_name.unique():\n",
    "    pos_wanted = list(out_tmp[out_tmp.gene_name == gene][\"pos\"])\n",
    "    pos_best = optimize_subsequence(pos_wanted, length=8, min_gap=40, better_gap=80, gene=gene)\n",
    "    pos_output = out_tmp[out_tmp.gene_name == gene]\n",
    "    pos_output = pos_output[pos_output[\"pos\"].isin(pos_best)]\n",
    "    output_df = pd.concat([output_df, pos_output])\n",
    "\n",
    "# write the output to a xlsx file\n",
    "output_df.to_csv(os.path.join(output, \"probes_wanted.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d783b3-7ed5-4f33-9673-9bb360c0002d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 经过上述步骤 已经生成想要的probes 存放在probes_wanted.csv文件中\n",
    "# 需要把输出的probes_wanted.csv放在example_dataset文件夹下面"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "16c1d0cb-2f47-40e9-83a7-1d588d367018",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "probe/example_dataset/results\n"
     ]
    }
   ],
   "source": [
    "resultsdir = os.path.join(workdir, 'results')\n",
    "print(resultsdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "47c983d1-fdb2-4d60-b0f4-e5dd07eb2f4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>organism</th>\n",
       "      <th>pos</th>\n",
       "      <th>plp_bds</th>\n",
       "      <th>plp_Tm</th>\n",
       "      <th>plp_bds3'</th>\n",
       "      <th>plp_bds5'</th>\n",
       "      <th>plp_Tm3'</th>\n",
       "      <th>plp_Tm5'</th>\n",
       "      <th>mfe</th>\n",
       "      <th>wanted</th>\n",
       "      <th>align_num</th>\n",
       "      <th>align_accession</th>\n",
       "      <th>align_descrip</th>\n",
       "      <th>plus/minus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>59</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG</td>\n",
       "      <td>72.43</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGG</td>\n",
       "      <td>GAGCATCGTCGCCCGCGAAG</td>\n",
       "      <td>57.44</td>\n",
       "      <td>57.81</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>60</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA</td>\n",
       "      <td>72.24</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGG</td>\n",
       "      <td>GGAGCATCGTCGCCCGCGAA</td>\n",
       "      <td>55.15</td>\n",
       "      <td>59.78</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>112</td>\n",
       "      <td>TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG</td>\n",
       "      <td>75.40</td>\n",
       "      <td>TCCTTCTGACCCATTCCCAC</td>\n",
       "      <td>CATCACACCCTGGTGCCTAG</td>\n",
       "      <td>59.80</td>\n",
       "      <td>58.05</td>\n",
       "      <td>-7.9</td>\n",
       "      <td>True</td>\n",
       "      <td>96</td>\n",
       "      <td>AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>120</td>\n",
       "      <td>CATAGGAGTCCTTCTGACCCATTCCCACCATCACACCCTG</td>\n",
       "      <td>72.19</td>\n",
       "      <td>CATAGGAGTCCTTCTGACCC</td>\n",
       "      <td>ATTCCCACCATCACACCCTG</td>\n",
       "      <td>55.55</td>\n",
       "      <td>58.10</td>\n",
       "      <td>-6.7</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>124</td>\n",
       "      <td>CCCACATAGGAGTCCTTCTGACCCATTCCCACCATCACAC</td>\n",
       "      <td>72.15</td>\n",
       "      <td>CCCACATAGGAGTCCTTCTG</td>\n",
       "      <td>ACCCATTCCCACCATCACAC</td>\n",
       "      <td>55.53</td>\n",
       "      <td>58.04</td>\n",
       "      <td>-6.7</td>\n",
       "      <td>True</td>\n",
       "      <td>47</td>\n",
       "      <td>AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession gene_name        mol_type organism  pos  \\\n",
       "10  ENSMUST00000163829      Actb  protein_coding    mouse   59   \n",
       "11  ENSMUST00000163829      Actb  protein_coding    mouse   60   \n",
       "12  ENSMUST00000163829      Actb  protein_coding    mouse  112   \n",
       "13  ENSMUST00000163829      Actb  protein_coding    mouse  120   \n",
       "14  ENSMUST00000163829      Actb  protein_coding    mouse  124   \n",
       "\n",
       "                                     plp_bds  plp_Tm             plp_bds3'  \\\n",
       "10  GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG   72.43  GGAGGGGAATACAGCCCGGG   \n",
       "11  TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA   72.24  TGGAGGGGAATACAGCCCGG   \n",
       "12  TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG   75.40  TCCTTCTGACCCATTCCCAC   \n",
       "13  CATAGGAGTCCTTCTGACCCATTCCCACCATCACACCCTG   72.19  CATAGGAGTCCTTCTGACCC   \n",
       "14  CCCACATAGGAGTCCTTCTGACCCATTCCCACCATCACAC   72.15  CCCACATAGGAGTCCTTCTG   \n",
       "\n",
       "               plp_bds5'  plp_Tm3'  plp_Tm5'  mfe  wanted  align_num  \\\n",
       "10  GAGCATCGTCGCCCGCGAAG     57.44     57.81 -8.4    True         43   \n",
       "11  GGAGCATCGTCGCCCGCGAA     55.15     59.78 -8.4    True         43   \n",
       "12  CATCACACCCTGGTGCCTAG     59.80     58.05 -7.9    True         96   \n",
       "13  ATTCCCACCATCACACCCTG     55.55     58.10 -6.7    True         55   \n",
       "14  ACCCATTCCCACCATCACAC     55.53     58.04 -6.7    True         47   \n",
       "\n",
       "                                      align_accession  \\\n",
       "10  AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...   \n",
       "11  AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...   \n",
       "12  AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...   \n",
       "13  AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...   \n",
       "14  AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...   \n",
       "\n",
       "                                        align_descrip  \\\n",
       "10   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "11   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "12   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "13   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "14   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "\n",
       "                                           plus/minus  \n",
       "10  -1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...  \n",
       "11  -1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...  \n",
       "12  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...  \n",
       "13  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...  \n",
       "14  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 去掉重复的probe\n",
    "resultsdir = os.path.join(workdir, 'results')\n",
    "def adjust_gene_name(gene_name, gene_list):\n",
    "    gene_list = [x.upper() for x in gene_list]\n",
    "    match = re.search(r'(.+)-(\\d+)$', gene_name)\n",
    "    if match:\n",
    "        base_gene_name = match.group(1)\n",
    "        if base_gene_name.upper() in gene_list or gene_name.upper() in gene_list: return base_gene_name\n",
    "        else: return gene_name\n",
    "    else: return gene_name\n",
    "\n",
    "\n",
    "result = pd.DataFrame()\n",
    "for dir in os.listdir(os.path.join(resultsdir)):\n",
    "    try: result = pd.concat([result, pd.read_csv(os.path.join(resultsdir, dir, \"probes_wanted.csv\"), index_col=0)])\n",
    "    except: continue\n",
    "result[\"gene_name\"] = [adjust_gene_name(_, gene_list) for _ in result[\"gene_name\"]]\n",
    "result = result.sort_values([\"gene_name\", \"pos\"])\n",
    "# result.drop_duplicates(subset=[\"bds\"], keep=\"first\", inplace=True)\n",
    "result.drop_duplicates(subset=[\"plp_bds\"], keep=\"first\", inplace=True)\n",
    "result.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "61c87abf-9f34-432d-8b5d-68d3f710c958",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('probe/example_dataset')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 注意查看目录，不然结果输出了也不知道放在哪里\n",
    "workdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4c3076b4-9de4-4291-a1a3-93b8bb3a00b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 保存结果\n",
    "result.to_csv(os.path.join(workdir, 'gene_binding_site.csv'))\n",
    "\n",
    "result[\"gene_name\"] = result[\"gene_name\"].str.upper()\n",
    "to_search = [gene for gene in gene_list if len(result[result[\"gene_name\"] == gene.upper()]) < 1]\n",
    "with open(os.path.join(workdir, \"to_search.txt\"), \"w\") as f:\n",
    "    for line in to_search: f.write(line + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c67f0eb-2648-4c0f-a8a3-09e1415737b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# 到目前为止已经获取了gene_binding_site 后续可能需要将这个文件和barcode联合起来\n",
    "# \n",
    "# 下一步就是使用随机数，获取barcode了"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4bd70b7a-632c-4754-9720-7267597acc72",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 导入模块和作者自己设置的函数\n",
    "from seqfold import dg, fold, dot_bracket\n",
    "import random\n",
    "\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "def random_seq_list(length=20, num=50):\n",
    "    nucleotides = ['A', 'T', 'C', 'G']\n",
    "    sequence = [''.join(random.choice(nucleotides) for _ in range(length)) for j in range(num)]\n",
    "    return sequence\n",
    "\n",
    "\n",
    "def hum_dis(seq1, seq2):\n",
    "    if len(seq1) != len(seq2):\n",
    "        print(\"seq_not_match\")\n",
    "    else:\n",
    "        cont = 0\n",
    "        for char in range(len(seq1)):\n",
    "            if seq1[char] != seq2[char]:\n",
    "                cont += 1\n",
    "        return cont\n",
    "\n",
    "\n",
    "def dna_sec_struct(seq, temp=45):\n",
    "    # Predict the minimum free energy\n",
    "    mfe = dg(seq, temp=temp)\n",
    "    # `fold` returns a list of `seqfold.Struct` from the minimum free energy structure\n",
    "    structs = fold(seq, temp=temp)\n",
    "    return mfe, structs\n",
    "\n",
    "\n",
    "def thre_by_blast(file=\"./JXAKR9US016-Alignment.xml\", thre=18):\n",
    "    pos = []\n",
    "    with open(file, \"r\") as blast_output:\n",
    "        blast_records = NCBIXML.parse(blast_output)\n",
    "        for blast_record in blast_records:\n",
    "            save = True\n",
    "            for alignment in blast_record.alignments:\n",
    "                # print(\"Alignment title:\", alignment.title)\n",
    "                # print(\"Length of the alignment:\", alignment.length)\n",
    "\n",
    "                # # Iterate over the high-scoring pairs (HSPs) in the alignment\n",
    "                for hsp in alignment.hsps:\n",
    "                    # print(\"HSP score:\", hsp.score)\n",
    "                    if hsp.score >= thre:\n",
    "                        pos.append(False)\n",
    "                        save = False\n",
    "                        break\n",
    "                    # print(\"HSP bits:\", hsp.bits)\n",
    "                    # print(\"HSP query sequence:\", hsp.query)\n",
    "                    # print(\"HSP match sequence:\", hsp.match)\n",
    "                    # print(\"HSP subject sequence:\", hsp.sbjct)\n",
    "            if save:\n",
    "                pos.append(True)\n",
    "    return pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7d8d372-fb5c-44a1-a643-79cf2dc797f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 使用上面的函数，生成符合条件的序列"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ea72c89-78d0-4a1e-9260-66819d3de6ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 注意下面生成随机序列 可以重复 注意文件保存"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "580f3de2-da22-4eae-8a1c-313fc774c01d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 注意下面生成随机序列 可以重复 注意文件保存"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16c0f3a9-f7bc-4fe8-a447-2b04dc39bcbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 注意下面生成随机序列 可以重复 注意文件保存"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7195a634-387b-4d19-a288-68b13501fbab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCCCACATGGAGTCCCAACT: \tthre_by_stru\t (((.....))).........\n",
      "GCCTCCTAGCAGGGATGCTG: \tthre_by_stru\t ((.(((......))).))..\n",
      "AGGACTCGTAGACCGACATT: \tthre_by_stru\t .((.((...)).))......\n",
      "ATGCCCATAGGTCAGTAAAG: \tthre_by_stru\t .((.((...)).))......\n",
      "CCTGTCGGTGTTAACAGGAT: \tthre_by_stru\t (((((........)))))..\n",
      "TCAATTGGGGGCGCGATGAT: \tthre_by_G\n",
      "TCCAGTTGTTTTGGGACGAA: \tthre_by_dif\n",
      "TCATATGTGCAAACAAGTGC: \tthre_by_dif\n",
      "GCTTCAGGCAAATCGCGGCC: \tthre_by_stru\t ......(((........)))\n",
      "CTCTGCGTTGCAGACGAACA: \tthre_by_stru\t .(((((...)))))......\n",
      "ACATCCGATTAGCATCGTCC: \tthre_by_stru\t .....((((....))))...\n",
      "TCTAACTGTCCGCAGACACC: \tthre_by_stru\t ......((((....))))..\n",
      "ACTTATATGGGGATACCAGA: \tthre_by_stru\t .........((....))...\n",
      "37 seqs remained\n"
     ]
    }
   ],
   "source": [
    "seq_list = random_seq_list()\n",
    "\n",
    "seq_list_export = []\n",
    "for seq in seq_list:\n",
    "    seq = seq.upper()\n",
    "\n",
    "    # GGGGG\n",
    "    if \"GGGGG\" in seq:\n",
    "        print(f\"{seq}: \\tthre_by_G\")\n",
    "        continue\n",
    "\n",
    "    # dif\n",
    "    dif = True\n",
    "    for tmp_seq in seq_list_export:\n",
    "        if hum_dis(tmp_seq, seq) < 10:\n",
    "            dif = False\n",
    "            print(f\"{seq}: \\tthre_by_dif\")\n",
    "    if not dif:\n",
    "        continue\n",
    "\n",
    "    # secondary structure\n",
    "    mfe, structs = dna_sec_struct(seq, temp=45)\n",
    "    if mfe < 0:\n",
    "        print(f\"{seq}: \\tthre_by_stru\\t\", dot_bracket(seq, structs))\n",
    "        continue\n",
    "\n",
    "    seq_list_export.append(seq)\n",
    "\n",
    "print(f'{len(seq_list_export)} seqs remained')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5bdd98e7-a531-49e4-bb77-bb121222eb2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/data/t180501/1_WeChat/RNA_FISH\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "wd=os.getcwd()\n",
    "print(wd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c41f4b5c-20f2-42af-903b-da047c124d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 保存文件,注意查看文件的输出位置\n",
    "with open(\"./random_seq_filtered.txt\", \"w\") as f:\n",
    "    for _ in range(len(seq_list_export)):\n",
    "        f.write(f'>seq{_}\\n' + seq_list_export[_] + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "017978fb-8248-46ed-a2b3-8365e8cd164c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 这个数据需要再次传递到ncbi 输出的xml文件有一项评分 需要筛选评分比较低的\n",
    "## https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=tblastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome\n",
    "## 这次就不用修改xml的文件名直接读取就好"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5b3c9588-592d-44ef-b73a-f51898787742",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 这个步骤主要是选出那些匹配程度较低的序列\n",
    "## 这样特意设计的序列估计是不不容易自己和自己形成环\n",
    "## 注意阈值的设置\n",
    "# pos = thre_by_blast(file='./1KDHC5BG013-Alignment.xml', thre=18)\n",
    "pos = thre_by_blast(file='./1KDHC5BG013-Alignment.xml', thre=20)\n",
    "sum(pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9df063cd-9a83-4c0f-b7f2-9fbf58123855",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 保存筛选之后的随机序列 \n",
    "with open(\"./random_seq_filtered.txt\", \"w\") as f:\n",
    "    cont = 0\n",
    "    for _ in range(len(seq_list_export)):\n",
    "        if pos[_]:\n",
    "            f.write(f\">seq{cont}\\n\" + seq_list_export[_] + \"\\n\")\n",
    "            cont += 1\n",
    "## 看一下随机序列\n",
    "tmp=pd.read_table('./random_seq_filtered.txt')\n",
    "# print(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3510afb0-8ee3-4971-9969-faebbaf58490",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 根据随机序列生成barcode序列 \n",
    "## 设置运行环境\n",
    "## 设置函数\n",
    "import pandas as pd\n",
    "\n",
    "def hum_dis(seq1, seq2):\n",
    "    seq1 = seq1.upper()\n",
    "    seq2 = seq2.upper()\n",
    "    if len(seq1) != len(seq2):\n",
    "        # print(\"seq_not_match\")\n",
    "        # print(seq1, seq2)\n",
    "        return -1\n",
    "    else:\n",
    "        cont = 0\n",
    "        for char in range(len(seq1)):\n",
    "            if seq1[char] != seq2[char]:\n",
    "                cont += 1\n",
    "        return cont\n",
    "    \n",
    "\n",
    "def create_seq_lib(\n",
    "    seq_list,\n",
    "    color_fraction={\n",
    "        \"Red\": [_ / 4 for _ in range(5)],\n",
    "        \"Green\": [_ / 2 for _ in range(3)],\n",
    "        \"Blue\": [_ / 4 for _ in range(5)],\n",
    "        \"Yellow\": [_ / 4 for _ in range(5)],\n",
    "    },\n",
    "):\n",
    "    seq_lib = pd.DataFrame(columns=[\"seq\", \"color\", \"grade\", \"fraction\"])\n",
    "    color_list = []\n",
    "    for color in color_fraction.keys():\n",
    "        color_list += [color] * len(color_fraction[color])\n",
    "    seq_lib[\"color\"] = color_list\n",
    "\n",
    "    if len(seq_lib[\"color\"]) > len(seq_list):\n",
    "        print(\"Seq Not Enough\")\n",
    "        return ValueError\n",
    "\n",
    "    seq_lib[\"seq\"] = seq_list[: len(seq_lib[\"color\"])]\n",
    "\n",
    "    fra = []\n",
    "    grade = []\n",
    "    for color in color_fraction.keys():\n",
    "        fra += color_fraction[color]\n",
    "        grade += [_ for _ in range(len(color_fraction[color]))]\n",
    "    seq_lib[\"fraction\"] = fra\n",
    "    seq_lib[\"grade\"] = grade\n",
    "\n",
    "    return seq_lib\n",
    "\n",
    "\n",
    "import itertools\n",
    "\n",
    "def create_barcode_lib(\n",
    "    seq_lib,\n",
    "    color_order=[\"Green\", \"Red\", \"Blue\", \"Yellow\"],\n",
    "    sum_num=5,\n",
    "    sum_list=[\"Red\", \"Yellow\", \"Blue\"],\n",
    "):\n",
    "    color_order_seq = [_ + \"seq\" for _ in color_order]\n",
    "    barcode_lib = pd.DataFrame(columns=[\"barcode\"] + color_order + color_order_seq)\n",
    "\n",
    "    grade_list = {\n",
    "        color: list(seq_lib[seq_lib.color == color].grade.unique())\n",
    "        for color in color_order\n",
    "    }\n",
    "    barcode_lib[color_order] = list(itertools.product(*grade_list.values()))\n",
    "\n",
    "    for i in range(len(barcode_lib)):\n",
    "        grade = barcode_lib.loc[i, color_order]\n",
    "\n",
    "        barcode_sub_list = [\n",
    "            list(\n",
    "                seq_lib[\n",
    "                    (seq_lib.color == color_order[_]) & (seq_lib.grade == grade[_])\n",
    "                ].seq\n",
    "            )[0]\n",
    "            for _ in range(len(color_order))\n",
    "        ]\n",
    "        barcode_lib.loc[i, color_order_seq] = barcode_sub_list\n",
    "        barcode_lib.loc[i, \"barcode\"] = \"\".join(barcode_sub_list)\n",
    "\n",
    "    if sum_num:\n",
    "        barcode_lib[\"sum\"] = barcode_lib[sum_list].sum(axis=1)\n",
    "        barcode_lib = barcode_lib[\n",
    "            # (barcode_lib[\"sum\"] >= 1) \n",
    "            # &\n",
    "            (barcode_lib[\"sum\"] == sum_num)\n",
    "        ]\n",
    "        barcode_lib.set_index('barcode', inplace=True)\n",
    "\n",
    "    return barcode_lib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "160463a5-520f-4ae0-84c4-f0cad8fe6fea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GGGGTTGAATGGGTAGATTC', 'CACTCTATCCACGGTTAAGA', 'TCCTCCTTTTTTGACACGCA', 'ACCTTACAATTTTTTACATC', 'AACAACGGTCTGGATGTTCA', 'TACTCGTGTTCCATCCAGTT', 'GACCATGCCCCCGTGTACAT', 'TCAAGCCAGGTCGATTATCT', 'ATGACTCCAATTTGTCACCC', 'ATGCTTCCTCATAGGGGCGA', 'CGTACAATAAGTGGCTTATC', 'GAGCCAGAACAGACCCTCGG', 'TTAGGTATACACAAGAGTTC', 'TAGCACACCGCATATGGGAA', 'TGTGCCCATAGGGTTAACCG', 'TTCACATCTCCTACCTGTGT', 'CGTGATGCTCTTATCAATCA', 'AGCTGCCCTCGCGTAACCCC', 'GCTGGTACGTACGCAACTTC', 'AGCAGACCCTGATCATCAAC', 'GCCTAACAGTTCCCACCCTT', 'CTCATGCCCTACAAATTCTG', 'CAGAGGTCGCCATCCCTTCA']\n"
     ]
    }
   ],
   "source": [
    "## 读取上个random_seq_filtered.txt文件\n",
    "file = './random_seq_filtered.txt'\n",
    "with open(file, \"r\") as f:\n",
    "    seq_list = f.readlines()\n",
    "seq_list = [seq_list[_].replace('\\n','') for _ in range(1,len(seq_list),2)]\n",
    "print(seq_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d9987a88-2d42-4526-86f0-0a2dd889bb3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>color</th>\n",
       "      <th>grade</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>Green</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CACTCTATCCACGGTTAAGA</td>\n",
       "      <td>Green</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCCTCCTTTTTTGACACGCA</td>\n",
       "      <td>Green</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>Red</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACAACGGTCTGGATGTTCA</td>\n",
       "      <td>Red</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TACTCGTGTTCCATCCAGTT</td>\n",
       "      <td>Red</td>\n",
       "      <td>2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GACCATGCCCCCGTGTACAT</td>\n",
       "      <td>Red</td>\n",
       "      <td>3</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCAAGCCAGGTCGATTATCT</td>\n",
       "      <td>Red</td>\n",
       "      <td>4</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ATGACTCCAATTTGTCACCC</td>\n",
       "      <td>Red</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ATGCTTCCTCATAGGGGCGA</td>\n",
       "      <td>Blue</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CGTACAATAAGTGGCTTATC</td>\n",
       "      <td>Blue</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GAGCCAGAACAGACCCTCGG</td>\n",
       "      <td>Blue</td>\n",
       "      <td>2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TTAGGTATACACAAGAGTTC</td>\n",
       "      <td>Blue</td>\n",
       "      <td>3</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TAGCACACCGCATATGGGAA</td>\n",
       "      <td>Blue</td>\n",
       "      <td>4</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TGTGCCCATAGGGTTAACCG</td>\n",
       "      <td>Blue</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTCACATCTCCTACCTGTGT</td>\n",
       "      <td>Yellow</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CGTGATGCTCTTATCAATCA</td>\n",
       "      <td>Yellow</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AGCTGCCCTCGCGTAACCCC</td>\n",
       "      <td>Yellow</td>\n",
       "      <td>2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GCTGGTACGTACGCAACTTC</td>\n",
       "      <td>Yellow</td>\n",
       "      <td>3</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AGCAGACCCTGATCATCAAC</td>\n",
       "      <td>Yellow</td>\n",
       "      <td>4</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GCCTAACAGTTCCCACCCTT</td>\n",
       "      <td>Yellow</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     seq   color  grade  fraction\n",
       "0   GGGGTTGAATGGGTAGATTC   Green      0       0.0\n",
       "1   CACTCTATCCACGGTTAAGA   Green      1       0.5\n",
       "2   TCCTCCTTTTTTGACACGCA   Green      2       1.0\n",
       "3   ACCTTACAATTTTTTACATC     Red      0       0.0\n",
       "4   AACAACGGTCTGGATGTTCA     Red      1       0.2\n",
       "5   TACTCGTGTTCCATCCAGTT     Red      2       0.4\n",
       "6   GACCATGCCCCCGTGTACAT     Red      3       0.6\n",
       "7   TCAAGCCAGGTCGATTATCT     Red      4       0.8\n",
       "8   ATGACTCCAATTTGTCACCC     Red      5       1.0\n",
       "9   ATGCTTCCTCATAGGGGCGA    Blue      0       0.0\n",
       "10  CGTACAATAAGTGGCTTATC    Blue      1       0.2\n",
       "11  GAGCCAGAACAGACCCTCGG    Blue      2       0.4\n",
       "12  TTAGGTATACACAAGAGTTC    Blue      3       0.6\n",
       "13  TAGCACACCGCATATGGGAA    Blue      4       0.8\n",
       "14  TGTGCCCATAGGGTTAACCG    Blue      5       1.0\n",
       "15  TTCACATCTCCTACCTGTGT  Yellow      0       0.0\n",
       "16  CGTGATGCTCTTATCAATCA  Yellow      1       0.2\n",
       "17  AGCTGCCCTCGCGTAACCCC  Yellow      2       0.4\n",
       "18  GCTGGTACGTACGCAACTTC  Yellow      3       0.6\n",
       "19  AGCAGACCCTGATCATCAAC  Yellow      4       0.8\n",
       "20  GCCTAACAGTTCCCACCCTT  Yellow      5       1.0"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 根据上面的序列生成矩阵\n",
    "## 在这个步骤里面 如果提供的随机的序列不够 会报错 因为这些序列分不过来\n",
    "## 但如果本身并不需要那么多荧光颜色的话 这边生成多少种类颜色的代码可以修改\n",
    "## 此代码可根据需要进行修改\n",
    "seq_lib = create_seq_lib(\n",
    "    seq_list=seq_list,\n",
    "    color_fraction={\n",
    "        \"Green\": [0, 0.5, 1],\n",
    "        \"Red\": [_ / 5 for _ in range(6)],\n",
    "        \"Blue\": [_ / 5 for _ in range(6)],\n",
    "        \"Yellow\": [_ / 5 for _ in range(6)],\n",
    "    },\n",
    ")\n",
    "seq_lib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "235b71c1-c0ca-47b5-9376-ca1bdb360980",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Green</th>\n",
       "      <th>Red</th>\n",
       "      <th>Blue</th>\n",
       "      <th>Yellow</th>\n",
       "      <th>Greenseq</th>\n",
       "      <th>Redseq</th>\n",
       "      <th>Blueseq</th>\n",
       "      <th>Yellowseq</th>\n",
       "      <th>sum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTTCCTCATAGGGGCGAGCCTAACAGTTCCCACCCTT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>ATGCTTCCTCATAGGGGCGA</td>\n",
       "      <td>GCCTAACAGTTCCCACCCTT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACAATAAGTGGCTTATCAGCAGACCCTGATCATCAAC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>CGTACAATAAGTGGCTTATC</td>\n",
       "      <td>AGCAGACCCTGATCATCAAC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCAGAACAGACCCTCGGGCTGGTACGTACGCAACTTC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>GAGCCAGAACAGACCCTCGG</td>\n",
       "      <td>GCTGGTACGTACGCAACTTC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGTATACACAAGAGTTCAGCTGCCCTCGCGTAACCCC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>TTAGGTATACACAAGAGTTC</td>\n",
       "      <td>AGCTGCCCTCGCGTAACCCC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCACACCGCATATGGGAACGTGATGCTCTTATCAATCA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>TAGCACACCGCATATGGGAA</td>\n",
       "      <td>CGTGATGCTCTTATCAATCA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATCGTACAATAAGTGGCTTATCCGTGATGCTCTTATCAATCA</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TCCTCCTTTTTTGACACGCA</td>\n",
       "      <td>GACCATGCCCCCGTGTACAT</td>\n",
       "      <td>CGTACAATAAGTGGCTTATC</td>\n",
       "      <td>CGTGATGCTCTTATCAATCA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATGAGCCAGAACAGACCCTCGGTTCACATCTCCTACCTGTGT</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>TCCTCCTTTTTTGACACGCA</td>\n",
       "      <td>GACCATGCCCCCGTGTACAT</td>\n",
       "      <td>GAGCCAGAACAGACCCTCGG</td>\n",
       "      <td>TTCACATCTCCTACCTGTGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTATGCTTCCTCATAGGGGCGACGTGATGCTCTTATCAATCA</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TCCTCCTTTTTTGACACGCA</td>\n",
       "      <td>TCAAGCCAGGTCGATTATCT</td>\n",
       "      <td>ATGCTTCCTCATAGGGGCGA</td>\n",
       "      <td>CGTGATGCTCTTATCAATCA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTCGTACAATAAGTGGCTTATCTTCACATCTCCTACCTGTGT</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TCCTCCTTTTTTGACACGCA</td>\n",
       "      <td>TCAAGCCAGGTCGATTATCT</td>\n",
       "      <td>CGTACAATAAGTGGCTTATC</td>\n",
       "      <td>TTCACATCTCCTACCTGTGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCTCCTTTTTTGACACGCAATGACTCCAATTTGTCACCCATGCTTCCTCATAGGGGCGATTCACATCTCCTACCTGTGT</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TCCTCCTTTTTTGACACGCA</td>\n",
       "      <td>ATGACTCCAATTTGTCACCC</td>\n",
       "      <td>ATGCTTCCTCATAGGGGCGA</td>\n",
       "      <td>TTCACATCTCCTACCTGTGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Green  Red  Blue  Yellow  \\\n",
       "barcode                                                                        \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTTC...      0    0     0       5   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACAA...      0    0     1       4   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCAG...      0    0     2       3   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGTA...      0    0     3       2   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCACA...      0    0     4       1   \n",
       "...                                                   ...  ...   ...     ...   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATCGTACAA...      2    3     1       1   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATGAGCCAG...      2    3     2       0   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTATGCTTC...      2    4     0       1   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTCGTACAA...      2    4     1       0   \n",
       "TCCTCCTTTTTTGACACGCAATGACTCCAATTTGTCACCCATGCTTC...      2    5     0       0   \n",
       "\n",
       "                                                                Greenseq  \\\n",
       "barcode                                                                    \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTTC...  GGGGTTGAATGGGTAGATTC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACAA...  GGGGTTGAATGGGTAGATTC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCAG...  GGGGTTGAATGGGTAGATTC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGTA...  GGGGTTGAATGGGTAGATTC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCACA...  GGGGTTGAATGGGTAGATTC   \n",
       "...                                                                  ...   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATCGTACAA...  TCCTCCTTTTTTGACACGCA   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATGAGCCAG...  TCCTCCTTTTTTGACACGCA   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTATGCTTC...  TCCTCCTTTTTTGACACGCA   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTCGTACAA...  TCCTCCTTTTTTGACACGCA   \n",
       "TCCTCCTTTTTTGACACGCAATGACTCCAATTTGTCACCCATGCTTC...  TCCTCCTTTTTTGACACGCA   \n",
       "\n",
       "                                                                  Redseq  \\\n",
       "barcode                                                                    \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTTC...  ACCTTACAATTTTTTACATC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACAA...  ACCTTACAATTTTTTACATC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCAG...  ACCTTACAATTTTTTACATC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGTA...  ACCTTACAATTTTTTACATC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCACA...  ACCTTACAATTTTTTACATC   \n",
       "...                                                                  ...   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATCGTACAA...  GACCATGCCCCCGTGTACAT   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATGAGCCAG...  GACCATGCCCCCGTGTACAT   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTATGCTTC...  TCAAGCCAGGTCGATTATCT   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTCGTACAA...  TCAAGCCAGGTCGATTATCT   \n",
       "TCCTCCTTTTTTGACACGCAATGACTCCAATTTGTCACCCATGCTTC...  ATGACTCCAATTTGTCACCC   \n",
       "\n",
       "                                                                 Blueseq  \\\n",
       "barcode                                                                    \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTTC...  ATGCTTCCTCATAGGGGCGA   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACAA...  CGTACAATAAGTGGCTTATC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCAG...  GAGCCAGAACAGACCCTCGG   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGTA...  TTAGGTATACACAAGAGTTC   \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCACA...  TAGCACACCGCATATGGGAA   \n",
       "...                                                                  ...   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATCGTACAA...  CGTACAATAAGTGGCTTATC   \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATGAGCCAG...  GAGCCAGAACAGACCCTCGG   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTATGCTTC...  ATGCTTCCTCATAGGGGCGA   \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTCGTACAA...  CGTACAATAAGTGGCTTATC   \n",
       "TCCTCCTTTTTTGACACGCAATGACTCCAATTTGTCACCCATGCTTC...  ATGCTTCCTCATAGGGGCGA   \n",
       "\n",
       "                                                               Yellowseq  sum  \n",
       "barcode                                                                        \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTTC...  GCCTAACAGTTCCCACCCTT    5  \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACAA...  AGCAGACCCTGATCATCAAC    5  \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCAG...  GCTGGTACGTACGCAACTTC    5  \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGTA...  AGCTGCCCTCGCGTAACCCC    5  \n",
       "GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCACA...  CGTGATGCTCTTATCAATCA    5  \n",
       "...                                                                  ...  ...  \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATCGTACAA...  CGTGATGCTCTTATCAATCA    5  \n",
       "TCCTCCTTTTTTGACACGCAGACCATGCCCCCGTGTACATGAGCCAG...  TTCACATCTCCTACCTGTGT    5  \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTATGCTTC...  CGTGATGCTCTTATCAATCA    5  \n",
       "TCCTCCTTTTTTGACACGCATCAAGCCAGGTCGATTATCTCGTACAA...  TTCACATCTCCTACCTGTGT    5  \n",
       "TCCTCCTTTTTTGACACGCAATGACTCCAATTTGTCACCCATGCTTC...  TTCACATCTCCTACCTGTGT    5  \n",
       "\n",
       "[63 rows x 9 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 生成整合后的barcode\n",
    "## 这个步骤有个看不懂的警告\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "barcode_lib = create_barcode_lib(\n",
    "    seq_lib,\n",
    "    color_order=[\"Green\", \"Red\", \"Blue\", \"Yellow\"],\n",
    "    sum_num=5,\n",
    "    sum_list=[\"Red\", \"Yellow\", \"Blue\"],\n",
    ")\n",
    "barcode_lib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "15e1a2e5-64ac-44cd-aacf-bec2835ad2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 保存seq 保存生成的barcode\n",
    "#tmp=os.getcwd()\n",
    "seq_lib.to_csv(r\"./seq_lib.csv\")\n",
    "barcode_lib.to_csv(r\"./barcode_lib.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43137e16-9278-49de-a160-d3624939a688",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 到目前 probe 和 barcode 都有了\n",
    "## 这个 probe似乎还是添加了stich的probe \n",
    "## 所以下个步骤要将这两个部分粘合 文件分别是gene_binding_site.csv 和 barcode_lib.csv\n",
    "## 是什么文件和什么文件粘合 ???"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8bbe94e5-9f9d-4087-a89c-7d410799afbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>accession</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>organism</th>\n",
       "      <th>pos</th>\n",
       "      <th>plp_bds</th>\n",
       "      <th>plp_Tm</th>\n",
       "      <th>plp_bds3'</th>\n",
       "      <th>plp_bds5'</th>\n",
       "      <th>plp_Tm3'</th>\n",
       "      <th>plp_Tm5'</th>\n",
       "      <th>mfe</th>\n",
       "      <th>wanted</th>\n",
       "      <th>align_num</th>\n",
       "      <th>align_accession</th>\n",
       "      <th>align_descrip</th>\n",
       "      <th>plus/minus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>59</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG</td>\n",
       "      <td>72.43</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGG</td>\n",
       "      <td>GAGCATCGTCGCCCGCGAAG</td>\n",
       "      <td>57.44</td>\n",
       "      <td>57.81</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>60</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA</td>\n",
       "      <td>72.24</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGG</td>\n",
       "      <td>GGAGCATCGTCGCCCGCGAA</td>\n",
       "      <td>55.15</td>\n",
       "      <td>59.78</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "      <td>AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12</td>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>112</td>\n",
       "      <td>TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG</td>\n",
       "      <td>75.40</td>\n",
       "      <td>TCCTTCTGACCCATTCCCAC</td>\n",
       "      <td>CATCACACCCTGGTGCCTAG</td>\n",
       "      <td>59.80</td>\n",
       "      <td>58.05</td>\n",
       "      <td>-7.9</td>\n",
       "      <td>True</td>\n",
       "      <td>96</td>\n",
       "      <td>AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>120</td>\n",
       "      <td>CATAGGAGTCCTTCTGACCCATTCCCACCATCACACCCTG</td>\n",
       "      <td>72.19</td>\n",
       "      <td>CATAGGAGTCCTTCTGACCC</td>\n",
       "      <td>ATTCCCACCATCACACCCTG</td>\n",
       "      <td>55.55</td>\n",
       "      <td>58.10</td>\n",
       "      <td>-6.7</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>ENSMUST00000163829</td>\n",
       "      <td>Actb</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mouse</td>\n",
       "      <td>124</td>\n",
       "      <td>CCCACATAGGAGTCCTTCTGACCCATTCCCACCATCACAC</td>\n",
       "      <td>72.15</td>\n",
       "      <td>CCCACATAGGAGTCCTTCTG</td>\n",
       "      <td>ACCCATTCCCACCATCACAC</td>\n",
       "      <td>55.53</td>\n",
       "      <td>58.04</td>\n",
       "      <td>-6.7</td>\n",
       "      <td>True</td>\n",
       "      <td>47</td>\n",
       "      <td>AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...</td>\n",
       "      <td>Mus musculus mammary gland RCB-0526 Jyg-MC(A)...</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0           accession gene_name        mol_type organism  pos  \\\n",
       "0          10  ENSMUST00000163829      Actb  protein_coding    mouse   59   \n",
       "1          11  ENSMUST00000163829      Actb  protein_coding    mouse   60   \n",
       "2          12  ENSMUST00000163829      Actb  protein_coding    mouse  112   \n",
       "3          13  ENSMUST00000163829      Actb  protein_coding    mouse  120   \n",
       "4          14  ENSMUST00000163829      Actb  protein_coding    mouse  124   \n",
       "\n",
       "                                    plp_bds  plp_Tm             plp_bds3'  \\\n",
       "0  GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG   72.43  GGAGGGGAATACAGCCCGGG   \n",
       "1  TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA   72.24  TGGAGGGGAATACAGCCCGG   \n",
       "2  TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG   75.40  TCCTTCTGACCCATTCCCAC   \n",
       "3  CATAGGAGTCCTTCTGACCCATTCCCACCATCACACCCTG   72.19  CATAGGAGTCCTTCTGACCC   \n",
       "4  CCCACATAGGAGTCCTTCTGACCCATTCCCACCATCACAC   72.15  CCCACATAGGAGTCCTTCTG   \n",
       "\n",
       "              plp_bds5'  plp_Tm3'  plp_Tm5'  mfe  wanted  align_num  \\\n",
       "0  GAGCATCGTCGCCCGCGAAG     57.44     57.81 -8.4    True         43   \n",
       "1  GGAGCATCGTCGCCCGCGAA     55.15     59.78 -8.4    True         43   \n",
       "2  CATCACACCCTGGTGCCTAG     59.80     58.05 -7.9    True         96   \n",
       "3  ATTCCCACCATCACACCCTG     55.55     58.10 -6.7    True         55   \n",
       "4  ACCCATTCCCACCATCACAC     55.53     58.04 -6.7    True         47   \n",
       "\n",
       "                                     align_accession  \\\n",
       "0  AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...   \n",
       "1  AK166349.1|AK152844.1|AK078935.1|AK150866.1|NG...   \n",
       "2  AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...   \n",
       "3  AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...   \n",
       "4  AK166349.1|AK152844.1|AK150662.1|AK078935.1|AK...   \n",
       "\n",
       "                                       align_descrip  \\\n",
       "0   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "1   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "2   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "3   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "4   Mus musculus mammary gland RCB-0526 Jyg-MC(A)...   \n",
       "\n",
       "                                          plus/minus  \n",
       "0  -1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...  \n",
       "1  -1,-1,-1,-1,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1...  \n",
       "2  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...  \n",
       "3  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...  \n",
       "4  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 读取 gene_binding_site.csv文件\n",
    "binding_df = pd.read_csv(DATASET_DIR / RUNID / \"gene_binding_site.csv\")\n",
    "if organism == 'mouse': binding_df['gene_name'] = binding_df['gene_name'].str.capitalize()\n",
    "elif organism == 'human': binding_df['gene_name'] = binding_df['gene_name'].str.upper()\n",
    "# binding_df = binding_df[binding_df['gene_name'].isin(gene_list)]\n",
    "print(len(binding_df))\n",
    "binding_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "cf751b7a-25e0-41cf-a44d-4f4b1b85f78f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>Green</th>\n",
       "      <th>Red</th>\n",
       "      <th>Blue</th>\n",
       "      <th>Yellow</th>\n",
       "      <th>Greenseq</th>\n",
       "      <th>Redseq</th>\n",
       "      <th>Blueseq</th>\n",
       "      <th>Yellowseq</th>\n",
       "      <th>sum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTT...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>ATGCTTCCTCATAGGGGCGA</td>\n",
       "      <td>GCCTAACAGTTCCCACCCTT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>CGTACAATAAGTGGCTTATC</td>\n",
       "      <td>AGCAGACCCTGATCATCAAC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>GAGCCAGAACAGACCCTCGG</td>\n",
       "      <td>GCTGGTACGTACGCAACTTC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>TTAGGTATACACAAGAGTTC</td>\n",
       "      <td>AGCTGCCCTCGCGTAACCCC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCAC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTC</td>\n",
       "      <td>ACCTTACAATTTTTTACATC</td>\n",
       "      <td>TAGCACACCGCATATGGGAA</td>\n",
       "      <td>CGTGATGCTCTTATCAATCA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             barcode  Green  Red  Blue  \\\n",
       "0  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCATGCTT...      0    0     0   \n",
       "1  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...      0    0     1   \n",
       "2  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...      0    0     2   \n",
       "3  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...      0    0     3   \n",
       "4  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTAGCAC...      0    0     4   \n",
       "\n",
       "   Yellow              Greenseq                Redseq               Blueseq  \\\n",
       "0       5  GGGGTTGAATGGGTAGATTC  ACCTTACAATTTTTTACATC  ATGCTTCCTCATAGGGGCGA   \n",
       "1       4  GGGGTTGAATGGGTAGATTC  ACCTTACAATTTTTTACATC  CGTACAATAAGTGGCTTATC   \n",
       "2       3  GGGGTTGAATGGGTAGATTC  ACCTTACAATTTTTTACATC  GAGCCAGAACAGACCCTCGG   \n",
       "3       2  GGGGTTGAATGGGTAGATTC  ACCTTACAATTTTTTACATC  TTAGGTATACACAAGAGTTC   \n",
       "4       1  GGGGTTGAATGGGTAGATTC  ACCTTACAATTTTTTACATC  TAGCACACCGCATATGGGAA   \n",
       "\n",
       "              Yellowseq  sum  \n",
       "0  GCCTAACAGTTCCCACCCTT    5  \n",
       "1  AGCAGACCCTGATCATCAAC    5  \n",
       "2  GCTGGTACGTACGCAACTTC    5  \n",
       "3  AGCTGCCCTCGCGTAACCCC    5  \n",
       "4  CGTGATGCTCTTATCAATCA    5  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 告诉计算机这次用是那篇文章涉及的分析 \n",
    "## 其实可以修改这个if函数 只要最终输出的是barcode_df 这样一个矩阵\n",
    "## 先尝试将barcode列选出来\n",
    "PANEL = 'PRISM'\n",
    "if PANEL == 'PRISM':\n",
    "    probe_df = pd.DataFrame()\n",
    "    barcode_df = pd.read_csv(\"./barcode_lib.csv\")\n",
    "elif PANEL == 'SPRINTseq':\n",
    "    barcode_df = pd.read_excel(DATASET_DIR / \"SPRINTSEQ_369_barcode.xlsx\", index_col=0)[['Barcode sequence']]\n",
    "    primer_l = 'TCCCTACACGACGCTCTTCCGATCT'\n",
    "    primer_r = 'CATTCCTGCTGAACCGCTCTTCCGA'\n",
    "    barcode_df['Barcode(70bp)'] = primer_l + barcode_df['Barcode sequence'] + primer_r + barcode_df['Barcode sequence']\n",
    "barcode_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df301859-2708-4084-850d-d59f23e61a4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 接下来就是探针粘贴的过程了\n",
    "## 可以直接进行粘贴 direct combine of binding site all\n",
    "## 也可以选择其中其中进行粘贴 select middle one for multi binding sites\n",
    "## 给每个基因选择3个备用选项 select 3 binding sites for each gene\n",
    "## probe_stitch.ipynb 是以下代码的来源  \n",
    "## 选择了以上的第3种方式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "407758b5-af41-43cd-9f4f-a20ac1da6ed9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PRISM</th>\n",
       "      <th>gene</th>\n",
       "      <th>probe_name</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>binding_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_1</td>\n",
       "      <td>PR_1_Actb_1</td>\n",
       "      <td>gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_2</td>\n",
       "      <td>PR_1_Actb_2</td>\n",
       "      <td>ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_3</td>\n",
       "      <td>PR_1_Actb_3</td>\n",
       "      <td>catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_1</td>\n",
       "      <td>PR_2_Gapdh_1</td>\n",
       "      <td>ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>CATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_2</td>\n",
       "      <td>PR_2_Gapdh_2</td>\n",
       "      <td>gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>CCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_3</td>\n",
       "      <td>PR_2_Gapdh_3</td>\n",
       "      <td>tgtgccgttgaatttgccgtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>TTCCCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Hprt1_1</td>\n",
       "      <td>PR_3_Hprt1_1</td>\n",
       "      <td>gtccataatcagtccatgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>CGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Hprt1_2</td>\n",
       "      <td>PR_3_Hprt1_2</td>\n",
       "      <td>tgtccataatcagtccatgaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>TCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Hprt1_3</td>\n",
       "      <td>PR_3_Hprt1_3</td>\n",
       "      <td>ctgtccataatcagtccatgGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>CTCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     PRISM     gene    probe_name  \\\n",
       "0  PRISM_1   Actb_1   PR_1_Actb_1   \n",
       "1  PRISM_1   Actb_2   PR_1_Actb_2   \n",
       "2  PRISM_1   Actb_3   PR_1_Actb_3   \n",
       "3  PRISM_2  Gapdh_1  PR_2_Gapdh_1   \n",
       "4  PRISM_2  Gapdh_2  PR_2_Gapdh_2   \n",
       "5  PRISM_2  Gapdh_3  PR_2_Gapdh_3   \n",
       "6  PRISM_3  Hprt1_1  PR_3_Hprt1_1   \n",
       "7  PRISM_3  Hprt1_2  PR_3_Hprt1_2   \n",
       "8  PRISM_3  Hprt1_3  PR_3_Hprt1_3   \n",
       "\n",
       "                                           probe_seq  \\\n",
       "0  gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "1  ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "2  catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "3  ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "4  gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "5  tgtgccgttgaatttgccgtGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "6  gtccataatcagtccatgagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "7  tgtccataatcagtccatgaGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "8  ctgtccataatcagtccatgGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "\n",
       "                                         barcode_seq  \\\n",
       "0  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "1  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "2  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "3  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "4  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "5  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "6  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...   \n",
       "7  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...   \n",
       "8  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...   \n",
       "\n",
       "                                binding_seq  \n",
       "0  GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG  \n",
       "1  TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA  \n",
       "2  TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG  \n",
       "3  CATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGT  \n",
       "4  CCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAG  \n",
       "5  TTCCCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGT  \n",
       "6  CGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGAG  \n",
       "7  TCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGA  \n",
       "8  CTCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATG  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 每个基因保存3条序列\n",
    "## 设置一些空参数\n",
    "probe_df = pd.DataFrame()\n",
    "cont = 0\n",
    "prism_pos = 0\n",
    "prism_pos_list = [_+1 for _ in range(9)]\n",
    "prism = prism_pos_list[prism_pos]\n",
    "max_cont = 3\n",
    "pre_gene_name = binding_df[\"gene_name\"].iloc[0]\n",
    "for num, gene in enumerate(binding_df[\"gene_name\"]):\n",
    "    if pre_gene_name != gene:\n",
    "        pre_gene_name = gene\n",
    "        cont = 0\n",
    "        prism_pos += 1\n",
    "        prism = prism_pos_list[prism_pos]\n",
    "    elif cont == max_cont:\n",
    "        continue\n",
    "    # print(num, gene, prism)\n",
    "    cont += 1\n",
    "    binding = binding_df[\"plp_bds\"].iloc[num]\n",
    "    assert len(binding) == 40, f\"binding site at pos {num} length is not 40bp: {binding}, {len(binding)} instead.\"\n",
    "\n",
    "    binding_l = binding[:20].lower()\n",
    "    binding_r = binding[20:].lower()\n",
    "    barcode = barcode_df.loc[prism, \"barcode\"]\n",
    "    probe = binding_r + barcode + binding_l\n",
    "\n",
    "    if PANEL == 'PRISM':\n",
    "        probe_info = pd.DataFrame({\n",
    "            \"PRISM\": [f\"PRISM_{prism}\"],\n",
    "            \"gene\":[f'{gene}_{cont}'],\n",
    "            \"probe_name\":[f'PR_{prism}_{gene}_{cont}'],\n",
    "            \"probe_seq\": [probe],\n",
    "            \"barcode_seq\": [barcode],\n",
    "            \"binding_seq\": [binding],})\n",
    "\n",
    "    elif PANEL == 'SPRINTseq':\n",
    "        probe_info = pd.DataFrame({\n",
    "            \"SPRINTseq\": [f\"SPRINTseq_{prism}\"],\n",
    "            \"gene\":[f'{gene}'],\n",
    "            \"probe_name\":[f'Seq_{prism}_{gene}_{cont}'],\n",
    "            \"probe\": [probe],\n",
    "            \"barcode\": [barcode],\n",
    "            \"binding\": [binding],})\n",
    "    if len(probe_df) == 0: probe_df = probe_info\n",
    "    else: probe_df = pd.concat([probe_df, probe_info])\n",
    "probe_df = probe_df.reset_index(drop=True)\n",
    "probe_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "6d95165e-2c06-46bb-a20a-6a785b99d192",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Glp1r_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glp1r_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glp1r_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Prkcd_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Prkcd_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Prkcd_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Vipr2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Vipr2_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Vipr2_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0\n",
       "0  Glp1r_1\n",
       "1  Glp1r_2\n",
       "2  Glp1r_3\n",
       "3  Prkcd_1\n",
       "4  Prkcd_2\n",
       "5  Prkcd_3\n",
       "6  Vipr2_1\n",
       "7  Vipr2_2\n",
       "8  Vipr2_3"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pd.DataFrame(probe_df['gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "fda84631-fea6-4656-909a-84e5d214feea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/data/t180501/3_zhangshiwen_RNA_FISH/FJ_New_EesyCode'"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import os\n",
    "# os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e41e0c9c-d5e0-407f-8a27-755b4610798d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PRISM</th>\n",
       "      <th>gene</th>\n",
       "      <th>probe_name</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>binding_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_1</td>\n",
       "      <td>PR_1_Actb_1</td>\n",
       "      <td>gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_2</td>\n",
       "      <td>PR_1_Actb_2</td>\n",
       "      <td>ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_3</td>\n",
       "      <td>PR_1_Actb_3</td>\n",
       "      <td>catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_1</td>\n",
       "      <td>PR_2_Gapdh_1</td>\n",
       "      <td>ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>CATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_2</td>\n",
       "      <td>PR_2_Gapdh_2</td>\n",
       "      <td>gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>CCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     PRISM     gene    probe_name  \\\n",
       "0  PRISM_1   Actb_1   PR_1_Actb_1   \n",
       "1  PRISM_1   Actb_2   PR_1_Actb_2   \n",
       "2  PRISM_1   Actb_3   PR_1_Actb_3   \n",
       "3  PRISM_2  Gapdh_1  PR_2_Gapdh_1   \n",
       "4  PRISM_2  Gapdh_2  PR_2_Gapdh_2   \n",
       "\n",
       "                                           probe_seq  \\\n",
       "0  gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "1  ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "2  catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "3  ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "4  gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "\n",
       "                                         barcode_seq  \\\n",
       "0  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "1  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "2  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "3  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "4  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "\n",
       "                                binding_seq  \n",
       "0  GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG  \n",
       "1  TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA  \n",
       "2  TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG  \n",
       "3  CATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGT  \n",
       "4  CCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAG  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probe_df.to_csv( './'+f'{PANEL}_probe.csv')\n",
    "print(len(probe_df))\n",
    "probe_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "b45c2e07-da1e-4b43-9761-726bd1b208fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 探针overlap检查\n",
    "# 探针和探针之间不要互相形成环"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "819b71f8-e135-48e5-8abf-5dd3a61946c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PRISM</th>\n",
       "      <th>gene</th>\n",
       "      <th>probe_name</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>binding_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_1</td>\n",
       "      <td>PR_1_Actb_1</td>\n",
       "      <td>gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_2</td>\n",
       "      <td>PR_1_Actb_2</td>\n",
       "      <td>ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Actb_3</td>\n",
       "      <td>PR_1_Actb_3</td>\n",
       "      <td>catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...</td>\n",
       "      <td>TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_1</td>\n",
       "      <td>PR_2_Gapdh_1</td>\n",
       "      <td>ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>CATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_2</td>\n",
       "      <td>PR_2_Gapdh_2</td>\n",
       "      <td>gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>CCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Gapdh_3</td>\n",
       "      <td>PR_2_Gapdh_3</td>\n",
       "      <td>tgtgccgttgaatttgccgtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...</td>\n",
       "      <td>TTCCCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Hprt1_1</td>\n",
       "      <td>PR_3_Hprt1_1</td>\n",
       "      <td>gtccataatcagtccatgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>CGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Hprt1_2</td>\n",
       "      <td>PR_3_Hprt1_2</td>\n",
       "      <td>tgtccataatcagtccatgaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>TCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Hprt1_3</td>\n",
       "      <td>PR_3_Hprt1_3</td>\n",
       "      <td>ctgtccataatcagtccatgGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "      <td>GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...</td>\n",
       "      <td>CTCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     PRISM     gene    probe_name  \\\n",
       "0  PRISM_1   Actb_1   PR_1_Actb_1   \n",
       "1  PRISM_1   Actb_2   PR_1_Actb_2   \n",
       "2  PRISM_1   Actb_3   PR_1_Actb_3   \n",
       "3  PRISM_2  Gapdh_1  PR_2_Gapdh_1   \n",
       "4  PRISM_2  Gapdh_2  PR_2_Gapdh_2   \n",
       "5  PRISM_2  Gapdh_3  PR_2_Gapdh_3   \n",
       "6  PRISM_3  Hprt1_1  PR_3_Hprt1_1   \n",
       "7  PRISM_3  Hprt1_2  PR_3_Hprt1_2   \n",
       "8  PRISM_3  Hprt1_3  PR_3_Hprt1_3   \n",
       "\n",
       "                                           probe_seq  \\\n",
       "0  gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "1  ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "2  catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "3  ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "4  gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "5  tgtgccgttgaatttgccgtGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "6  gtccataatcagtccatgagGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "7  tgtccataatcagtccatgaGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "8  ctgtccataatcagtccatgGGGGTTGAATGGGTAGATTCACCTTA...   \n",
       "\n",
       "                                         barcode_seq  \\\n",
       "0  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "1  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "2  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCCGTACA...   \n",
       "3  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "4  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "5  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCGAGCCA...   \n",
       "6  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...   \n",
       "7  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...   \n",
       "8  GGGGTTGAATGGGTAGATTCACCTTACAATTTTTTACATCTTAGGT...   \n",
       "\n",
       "                                binding_seq  \n",
       "0  GGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAAG  \n",
       "1  TGGAGGGGAATACAGCCCGGGGAGCATCGTCGCCCGCGAA  \n",
       "2  TCCTTCTGACCCATTCCCACCATCACACCCTGGTGCCTAG  \n",
       "3  CATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGT  \n",
       "4  CCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAG  \n",
       "5  TTCCCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGT  \n",
       "6  CGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGAG  \n",
       "7  TCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATGA  \n",
       "8  CTCGAGCAAGTCTTTCAGTCCTGTCCATAATCAGTCCATG  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 已经获取序列 \n",
    "# 对序列进行检查\n",
    "def overlap_degree(str1, str2):\n",
    "    max_overlap = min(len(str1), len(str2))\n",
    "    \n",
    "    for i in range(max_overlap, 0, -1):\n",
    "        if str1[-i:] == str2[:i]:\n",
    "            return i\n",
    "    \n",
    "    return 0\n",
    "meta = pd.read_csv(\"./PRISM_probe.csv\",index_col=0)\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "061018d0-b641-49d5-bcee-c98f44e2176d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PR_3_Vipr2_1, -1, 0, degree=36\n",
      "PR_3_Vipr2_1, 0, 1, degree=37\n",
      "PR_3_Vipr2_1, 1, 2, degree=39\n"
     ]
    }
   ],
   "source": [
    "# 计算序列存在重合的情况 避免形成圆圈\n",
    "# 输出的这些都是不能被使用的\n",
    "for prism in range(3): ## 多少个基因这边输入多少\n",
    "    detect_seq = list(meta[meta['PRISM'] == f'PRISM_{prism+1}']['binding_seq'])\n",
    "    gene = list(meta[meta['PRISM'] == f'PRISM_{prism+1}']['probe_name'])\n",
    "    gene = gene[0]\n",
    "    for i in range(-1,2,1):\n",
    "        degree = max(overlap_degree(detect_seq[i], detect_seq[i+1]), overlap_degree(detect_seq[i+1], detect_seq[i]))\n",
    "        if degree > 5:\n",
    "            print(f'{gene}, {i}, {i+1}, degree={degree}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "fdc08cd0-1620-4fd1-a4d6-d7d1a3889d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 接下来是空间位阻的检查"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "60fb47e7-dfe7-4528-bc9b-e8a5308a6185",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 又是先载入作者写的函数\n",
    "import pandas as pd\n",
    "def dna_sec_struct(seq, temp=45):\n",
    "    # Predict the minimum free energy\n",
    "    mfe = dg(seq, temp=temp)\n",
    "    # `fold` returns a list of `seqfold.Struct` from the minimum free energy structure\n",
    "    structs = fold(seq, temp=temp)\n",
    "    return mfe, structs\n",
    "\n",
    "def seq_minus(seq):\n",
    "    translib = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    return \"\".join(list(reversed([translib[i] for i in seq])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "1fe4927b-0566-4620-99ab-00f3756c6c3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PRISM</th>\n",
       "      <th>gene</th>\n",
       "      <th>probe_name</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>binding_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Glp1r_1</td>\n",
       "      <td>PR_1_Glp1r_1</td>\n",
       "      <td>acaccgtggtaccctgagggTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAACGGAT...</td>\n",
       "      <td>CTGCACCGTCTCTGAGAGGGACACCGTGGTACCCTGAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Glp1r_2</td>\n",
       "      <td>PR_1_Glp1r_2</td>\n",
       "      <td>aaggaacctgggggcccatcTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAACGGAT...</td>\n",
       "      <td>AGGGGCAGCTGACATTCACGAAGGAACCTGGGGGCCCATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRISM_1</td>\n",
       "      <td>Glp1r_3</td>\n",
       "      <td>PR_1_Glp1r_3</td>\n",
       "      <td>cacactccgacaggtccctcTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAACGGAT...</td>\n",
       "      <td>CTCCCCTCGCTTAGACTCTTCACACTCCGACAGGTCCCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Prkcd_1</td>\n",
       "      <td>PR_2_Prkcd_1</td>\n",
       "      <td>cacagaacagaaggtgggctTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAGATTCC...</td>\n",
       "      <td>CCCCAGACAAACTCTTTGCACACAGAACAGAAGGTGGGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Prkcd_2</td>\n",
       "      <td>PR_2_Prkcd_2</td>\n",
       "      <td>ggtcacagaaggtggggctcTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAGATTCC...</td>\n",
       "      <td>CCAGAGCAAACTGCCACAGTGGTCACAGAAGGTGGGGCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PRISM_2</td>\n",
       "      <td>Prkcd_3</td>\n",
       "      <td>PR_2_Prkcd_3</td>\n",
       "      <td>cttccgggaagatctctgggTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAGATTCC...</td>\n",
       "      <td>ACAGACTCTGTTGTGTCCAGCTTCCGGGAAGATCTCTGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Vipr2_1</td>\n",
       "      <td>PR_3_Vipr2_1</td>\n",
       "      <td>tgtggtcgtttgtgtcccagTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGACTATTT...</td>\n",
       "      <td>AATGACCCACCAGGGGATGCTGTGGTCGTTTGTGTCCCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Vipr2_2</td>\n",
       "      <td>PR_3_Vipr2_2</td>\n",
       "      <td>tgctgtggtcgtttgtgtccTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGACTATTT...</td>\n",
       "      <td>CCGAATGACCCACCAGGGGATGCTGTGGTCGTTTGTGTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PRISM_3</td>\n",
       "      <td>Vipr2_3</td>\n",
       "      <td>PR_3_Vipr2_3</td>\n",
       "      <td>atgctgtggtcgtttgtgtcTGCTATAATGAGTAGCTGAGTGACGC...</td>\n",
       "      <td>TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGACTATTT...</td>\n",
       "      <td>TCCGAATGACCCACCAGGGGATGCTGTGGTCGTTTGTGTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     PRISM     gene    probe_name  \\\n",
       "0  PRISM_1  Glp1r_1  PR_1_Glp1r_1   \n",
       "1  PRISM_1  Glp1r_2  PR_1_Glp1r_2   \n",
       "2  PRISM_1  Glp1r_3  PR_1_Glp1r_3   \n",
       "3  PRISM_2  Prkcd_1  PR_2_Prkcd_1   \n",
       "4  PRISM_2  Prkcd_2  PR_2_Prkcd_2   \n",
       "5  PRISM_2  Prkcd_3  PR_2_Prkcd_3   \n",
       "6  PRISM_3  Vipr2_1  PR_3_Vipr2_1   \n",
       "7  PRISM_3  Vipr2_2  PR_3_Vipr2_2   \n",
       "8  PRISM_3  Vipr2_3  PR_3_Vipr2_3   \n",
       "\n",
       "                                           probe_seq  \\\n",
       "0  acaccgtggtaccctgagggTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "1  aaggaacctgggggcccatcTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "2  cacactccgacaggtccctcTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "3  cacagaacagaaggtgggctTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "4  ggtcacagaaggtggggctcTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "5  cttccgggaagatctctgggTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "6  tgtggtcgtttgtgtcccagTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "7  tgctgtggtcgtttgtgtccTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "8  atgctgtggtcgtttgtgtcTGCTATAATGAGTAGCTGAGTGACGC...   \n",
       "\n",
       "                                         barcode_seq  \\\n",
       "0  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAACGGAT...   \n",
       "1  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAACGGAT...   \n",
       "2  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAACGGAT...   \n",
       "3  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAGATTCC...   \n",
       "4  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAGATTCC...   \n",
       "5  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGAGATTCC...   \n",
       "6  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGACTATTT...   \n",
       "7  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGACTATTT...   \n",
       "8  TGCTATAATGAGTAGCTGAGTGACGCCGCACGGTTCTCGACTATTT...   \n",
       "\n",
       "                                binding_seq  \n",
       "0  CTGCACCGTCTCTGAGAGGGACACCGTGGTACCCTGAGGG  \n",
       "1  AGGGGCAGCTGACATTCACGAAGGAACCTGGGGGCCCATC  \n",
       "2  CTCCCCTCGCTTAGACTCTTCACACTCCGACAGGTCCCTC  \n",
       "3  CCCCAGACAAACTCTTTGCACACAGAACAGAAGGTGGGCT  \n",
       "4  CCAGAGCAAACTGCCACAGTGGTCACAGAAGGTGGGGCTC  \n",
       "5  ACAGACTCTGTTGTGTCCAGCTTCCGGGAAGATCTCTGGG  \n",
       "6  AATGACCCACCAGGGGATGCTGTGGTCGTTTGTGTCCCAG  \n",
       "7  CCGAATGACCCACCAGGGGATGCTGTGGTCGTTTGTGTCC  \n",
       "8  TCCGAATGACCCACCAGGGGATGCTGTGGTCGTTTGTGTC  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pd.read_csv('./PRISM_probe.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fdc4b845-0a7b-4e3e-b292-18505fffc609",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe_name</th>\n",
       "      <th>probe_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PR_1_Actb_1</td>\n",
       "      <td>gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PR_1_Actb_2</td>\n",
       "      <td>ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PR_1_Actb_3</td>\n",
       "      <td>catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PR_2_Gapdh_1</td>\n",
       "      <td>ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PR_2_Gapdh_2</td>\n",
       "      <td>gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PR_2_Gapdh_3</td>\n",
       "      <td>tgtgccgttgaatttgccgtGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PR_3_Hprt1_1</td>\n",
       "      <td>gtccataatcagtccatgagGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PR_3_Hprt1_2</td>\n",
       "      <td>tgtccataatcagtccatgaGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PR_3_Hprt1_3</td>\n",
       "      <td>ctgtccataatcagtccatgGGGGTTGAATGGGTAGATTCACCTTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     probe_name                                          probe_seq\n",
       "0   PR_1_Actb_1  gagcatcgtcgcccgcgaagGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "1   PR_1_Actb_2  ggagcatcgtcgcccgcgaaGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "2   PR_1_Actb_3  catcacaccctggtgcctagGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "3  PR_2_Gapdh_1  ccgttgaatttgccgtgagtGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "4  PR_2_Gapdh_2  gccgttgaatttgccgtgagGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "5  PR_2_Gapdh_3  tgtgccgttgaatttgccgtGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "6  PR_3_Hprt1_1  gtccataatcagtccatgagGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "7  PR_3_Hprt1_2  tgtccataatcagtccatgaGGGGTTGAATGGGTAGATTCACCTTA...\n",
       "8  PR_3_Hprt1_3  ctgtccataatcagtccatgGGGGTTGAATGGGTAGATTCACCTTA..."
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 注意提供的序列并不是全部的序列，需要重新琢磨以下这个代码看这个具体的判断标准究竟是什么。\n",
    "# 可能要用binding seq也说不定\n",
    "# padlock = pd.read_excel(r'E:\\TMC\\probe_designer\\dataset\\2024.3.16_TCR&mutation_3_Breast_cancer\\binding_site_revised.xlsx', sheet_name='padlock')\n",
    "padlock = pd.read_csv('./PRISM_probe.csv',index_col=0)[['probe_name','probe_seq']]\n",
    "padlock"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "05f99db8-bd05-4a4f-9c53-cfcdd2864600",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     probe_name                                          probe_seq\n",
      "0   PR_1_Actb_1  GAGCATCGTCGCCCGCGAAGGGGGTTGAATGGGTAGATTCACCTTA...\n",
      "1   PR_1_Actb_2  GGAGCATCGTCGCCCGCGAAGGGGTTGAATGGGTAGATTCACCTTA...\n",
      "2   PR_1_Actb_3  CATCACACCCTGGTGCCTAGGGGGTTGAATGGGTAGATTCACCTTA...\n",
      "3  PR_2_Gapdh_1  CCGTTGAATTTGCCGTGAGTGGGGTTGAATGGGTAGATTCACCTTA...\n",
      "4  PR_2_Gapdh_2  GCCGTTGAATTTGCCGTGAGGGGGTTGAATGGGTAGATTCACCTTA...\n"
     ]
    }
   ],
   "source": [
    "## 将probe_seq列小写全部换成大写\n",
    "# 将 probe_seq 列中的小写字母转换为大写\n",
    "padlock[\"probe_seq\"] = padlock[\"probe_seq\"].str.upper()\n",
    "# 验证结果\n",
    "print(padlock.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "24b368e8-4ddc-4d5d-9abf-363ebb4f8b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 保存最终的文件\n",
    "padlock.to_csv('./probe_seq.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4de5428d-9302-4dcb-a133-e2960d1c3722",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PR_1_Actb_1:-0.3\n",
      "CCCGGGCTGTATTCCCCTCCGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTTCGCGGGCGACGATGCTCCCCGGGCTGTATTCCCCTCCGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTTCGCGGGCGACGATGCTCCCCGGGCTGTATTCCCCTCCGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTTCGCGGGCGACGATGCTC\n",
      ".....................................................................................................................................................................................................................................................................................(((....))).........................................................................\n",
      "PR_1_Actb_2:-0.3\n",
      "CCGGGCTGTATTCCCCTCCAGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCTTCGCGGGCGACGATGCTCCCCGGGCTGTATTCCCCTCCAGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCTTCGCGGGCGACGATGCTCCCCGGGCTGTATTCCCCTCCAGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCTTCGCGGGCGACGATGCTCC\n",
      ".....................................................................................................................................................................................................................................................................................(((....))).........................................................................\n",
      "PR_1_Actb_3:-1.6\n",
      "GTGGGAATGGGTCAGAAGGAGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTAGGCACCAGGGTGTGATGGTGGGAATGGGTCAGAAGGAGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTAGGCACCAGGGTGTGATGGTGGGAATGGGTCAGAAGGAGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTAGGCACCAGGGTGTGATG\n",
      "..............................................................................................................................................................................................................((((((((.(((((((.......))))......)))..))))))))............................................................................................................\n",
      "PR_2_Gapdh_1:-2.8\n",
      "CACAGTCAAGGCCGAGAATGGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCACTCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCACTCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCACTCACGGCAAATTCAACGG\n",
      "..........................................................................................................................................................................................................................(((.((.((((.....................)))).)).)))...................................................................................................\n",
      "PR_2_Gapdh_2:-1.4\n",
      "ACAGTCAAGGCCGAGAATGGGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTCACGGCAAATTCAACGGC\n",
      ".....................................................................................................................................................................................................................................................................................((((.((........)).)))).............................................................\n",
      "PR_2_Gapdh_3:-1.4\n",
      "GTCAAGGCCGAGAATGGGAAGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGGAAGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGGAAGAAGTTGCGTACGTACCAGCCCGAGGGTCTGTTCTGGCTCGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCACGGCAAATTCAACGGCACA\n",
      ".....................................................................................................................................................................................................................................................................................((((.((........)).)))).............................................................\n",
      "PR_3_Hprt1_1:-0.2\n",
      "AGGACTGAAAGACTTGCTCGGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTCATGGACTGATTATGGACAGGACTGAAAGACTTGCTCGGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTCATGGACTGATTATGGACAGGACTGAAAGACTTGCTCGGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTCATGGACTGATTATGGAC\n",
      "......................................................................................................................................................................................................................(((((((......................................)))))))..............................................................................................\n",
      "PR_3_Hprt1_2:-0.8\n",
      "GGACTGAAAGACTTGCTCGAGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCTCATGGACTGATTATGGACAGGACTGAAAGACTTGCTCGAGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCTCATGGACTGATTATGGACAGGACTGAAAGACTTGCTCGAGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCTCATGGACTGATTATGGACA\n",
      "......................................................................................................................................................................................................................((((((((....................................))))))))..............................................................................................\n",
      "PR_3_Hprt1_3:-0.6\n",
      "GACTGAAAGACTTGCTCGAGGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCATGGACTGATTATGGACAGGACTGAAAGACTTGCTCGAGGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCATGGACTGATTATGGACAGGACTGAAAGACTTGCTCGAGGGGGTTACGCGAGGGCAGCTGAACTCTTGTGTATACCTAAGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCATGGACTGATTATGGACAG\n",
      "......................................................................................................................................................................................................................(((((((......................................)))))))..............................................................................................\n"
     ]
    }
   ],
   "source": [
    "## 调用上面的函数 查看评分\n",
    "for index, (name, seq) in padlock.iterrows():\n",
    "    seq = seq_minus(seq)\n",
    "    seq = seq.upper() * 3\n",
    "    mfe, structs = dna_sec_struct(seq, temp=60)\n",
    "    # if mfe < 0:\n",
    "    print(f\"{name}:{mfe}\\n{seq}\\n{dot_bracket(seq, structs)}\")\n",
    "    # continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "3ddd1d8f-e179-4ac3-be82-692ae115c802",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "360\n"
     ]
    }
   ],
   "source": [
    "a='CCCGGGCTGTATTCCCCTCCGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTTCGCGGGCGACGATGCTCCCCGGGCTGTATTCCCCTCCGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTTCGCGGGCGACGATGCTCCCCGGGCTGTATTCCCCTCCGTTGATGATCAGGGTCTGCTGATAAGCCACTTATTGTACGGATGTAAAAAATTGTAAGGTGAATCTACCCATTCAACCCCCTTCGCGGGCGACGATGCTC'\n",
    "print(len(a))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Prism",
   "language": "python",
   "name": "prism"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
