{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe Designer\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "# get gene data from ncbi\n",
    "from Bio import Entrez\n",
    "\n",
    "# blast and xml file process\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "# dir\n",
    "os.chdir(\"/home/akikawa/repos/probe_designer/dataset/example\")\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "tmp = \"./results/\" + formatted_time + \"/tmp/\"\n",
    "output = \"./results/\" + formatted_time + \"/\"\n",
    "# tmp = './results/' + '20231026_203549' + '/tmp/'\n",
    "# output = './results/' + '20231026_203549' + '/'\n",
    "pre_binding_dir = tmp + \"pre_binding/\"\n",
    "try:\n",
    "    os.makedirs(tmp)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# basic variables\n",
    "gene_name_list_tosearch = \"gene_name_list_tosearch.txt\"\n",
    "pre_binding_file_suffix = \"_pre_binding.fasta\"\n",
    "total_pre_binding_file_name = \"_total_pre_binding.fasta\"\n",
    "cut_name = total_pre_binding_file_name.replace(\".fasta\", \"_cut.fasta\")\n",
    "\n",
    "# tmp file\n",
    "gene_name_list_file = \"1_gene_name_list.txt\"\n",
    "gene_id_name_file = \"2_id_list.txt\"\n",
    "gene_seq_in_file = \"3_gene_seq_in_file.gb\"\n",
    "pre_binding_num_file = \"4_pre_binding_num.json\"\n",
    "blast_results_file = \"5_blast_results.xml\"\n",
    "cut_blast_results_file = blast_results_file.replace(\".xml\", \"_cut.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initiation of array\n",
    "binding_site_FOIs = [\n",
    "    \"accession\",\n",
    "    \"gene_name\",\n",
    "    \"mol_type\",\n",
    "    \"organism\",\n",
    "    \"binding\",\n",
    "    \"Tm_l\",\n",
    "    \"Tm_r\",\n",
    "    \"wanted\",\n",
    "]\n",
    "align_FOIs = [\"align_num\", \"align_accession\", \"align_descrip\", \"plus/minus\"]\n",
    "FOI = pd.DataFrame(columns=binding_site_FOIs + align_FOIs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get genbank file of each gene from ncbi dataset\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get gene id and other information from ncbi dataset(api)\n",
    "## Generate gene_search_list from gene_name_list\n",
    "organism_of_interest = \"Homo sapiens\"\n",
    "n_type_of_interest = \"mRNA\"\n",
    "with open(tmp + gene_name_list_file) as f:\n",
    "    gene_name_list = f.read().splitlines()\n",
    "gene_search_list = [\n",
    "    \", \".join([name, organism_of_interest, n_type_of_interest])\n",
    "    for name in gene_name_list\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get gene id list using Entrez.esearch\n",
    "id_list = []\n",
    "for gene_search in gene_search_list:\n",
    "    Entrez.email = \"1418767067@qq.com\"\n",
    "    handle = Entrez.esearch(db=\"nuccore\", term=gene_search)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    id_list += record[\"IdList\"][:1]  # set number of search results to read\n",
    "with open(tmp + gene_id_name_file, \"w\") as f:\n",
    "    f.write(\"\\n\".join(id_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read id_list from existing file\n",
    "with open(tmp + gene_id_name_file, \"r\") as f:\n",
    "    id_list = f.read().split(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the genbank file of each gene by search for id list\n",
    "fetch_per_round = 3\n",
    "round = -(-len(id_list) // fetch_per_round)\n",
    "for i in range(round):\n",
    "    id_list_per_round = id_list[i * fetch_per_round : (i + 1) * fetch_per_round]\n",
    "    Entrez.email = \"1418767067@qq.com\"\n",
    "    handle = Entrez.efetch(\n",
    "        db=\"nuccore\",\n",
    "        strand=1,  # plus if strand=1\n",
    "        id=id_list_per_round,\n",
    "        rettype=\"gbwithparts\",\n",
    "        retmode=\"text\",\n",
    "    )\n",
    "    seq_record = handle.read()\n",
    "    handle.close()\n",
    "    print(i + 1, \"{:.1f} %\".format((i + 1) / round * 100))\n",
    "    with open(tmp + gene_seq_in_file, \"a\") as f:\n",
    "        f.write(seq_record)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binding site Searcher\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def select_random_non_overlapping_substrings(input_string, length, num_substrings):\n",
    "    if length > len(input_string) or num_substrings * length > len(input_string):\n",
    "    # if length > len(input_string):\n",
    "        raise ValueError(\"Invalid input parameters.\")\n",
    "\n",
    "    available_positions = list(range(len(input_string) - length + 1))\n",
    "    random.shuffle(available_positions)\n",
    "\n",
    "    substrings = []\n",
    "    positions = set()  # To keep track of selected positions\n",
    "\n",
    "    for _ in range(num_substrings):\n",
    "        if not available_positions:\n",
    "            break  # Stop if there are no more non-overlapping positions\n",
    "\n",
    "        # Try to find a non-overlapping position\n",
    "        selected_position = available_positions.pop()\n",
    "        while selected_position in positions:\n",
    "            if not available_positions:\n",
    "                break\n",
    "            selected_position = available_positions.pop()\n",
    "\n",
    "        if selected_position not in positions:\n",
    "            positions.add(selected_position)\n",
    "            end = selected_position + length\n",
    "            selected_substring = input_string[selected_position:end]\n",
    "            substrings.append(selected_substring)\n",
    "\n",
    "    return substrings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Search binding sites on mRNA sequence\n",
    "file_in = tmp + gene_seq_in_file\n",
    "file_out_dir = pre_binding_dir\n",
    "try:\n",
    "    os.mkdir(file_out_dir)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "BDS_len = 40\n",
    "Potential_len = 40\n",
    "max_num = 100\n",
    "\n",
    "pot_st, pot_en = BDS_len // 2 - Potential_len // 2, BDS_len // 2 + Potential_len // 2\n",
    "pre_binding_num = {}\n",
    "pos, length = 0, 0\n",
    "for gene_seq_in in SeqIO.parse(tmp + gene_seq_in_file, \"genbank\"):\n",
    "    # get information of gene\n",
    "    id = gene_seq_in.id  # get seq id\n",
    "    # get gene_name\n",
    "    if gene_seq_in.features:\n",
    "        for feature in gene_seq_in.features:\n",
    "            if feature.type == \"CDS\":\n",
    "                gene_name = feature.qualifiers.get(\"gene\", [\"NAN\"])[0]\n",
    "    # get molecule_type\n",
    "    mol_type = gene_seq_in.annotations[\"molecule_type\"]\n",
    "    # get organism\n",
    "    organism = gene_seq_in.annotations[\"organism\"]\n",
    "    # get minus seq\n",
    "    translib = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    try:\n",
    "        seq_minus = [translib[i] for i in str(gene_seq_in.seq)]\n",
    "        seq = \"\".join(list(reversed(seq_minus)))\n",
    "    except:\n",
    "        seq = str(gene_seq_in.seq)\n",
    "\n",
    "    # set start point and pre_binding_num\n",
    "    length = len(seq)\n",
    "    pre_binding_num_tmp = min(length // BDS_len, max_num)\n",
    "    st = length // 2 - pre_binding_num_tmp * BDS_len // 2\n",
    "    # generate and write pre_binding for each gene in a fasta file\n",
    "    record_list = []\n",
    "    file_out = file_out_dir + id + pre_binding_file_suffix\n",
    "\n",
    "    # for i in range(pre_binding_num_tmp):\n",
    "    #     pre_binding_tmp = seq[st + i * BDS_len : st + i * BDS_len + BDS_len]\n",
    "\n",
    "    selected_substrings = select_random_non_overlapping_substrings(\n",
    "        seq, BDS_len, pre_binding_num_tmp\n",
    "    )\n",
    "\n",
    "    for i, pre_binding_tmp in enumerate(selected_substrings):\n",
    "        Tm_l = mt.Tm_NN(pre_binding_tmp[: BDS_len // 2], nn_table=mt.R_DNA_NN1)\n",
    "        Tm_r = mt.Tm_NN(pre_binding_tmp[BDS_len // 2 :], nn_table=mt.R_DNA_NN1)\n",
    "        record_list.append(\n",
    "            SeqRecord(\n",
    "                Seq(pre_binding_tmp),\n",
    "                id=\"pre_binding\" + str(i),\n",
    "                description=\"|\".join([id, gene_name, organism, mol_type]),\n",
    "            )\n",
    "        )\n",
    "        # add information about binding sites to FOI\n",
    "        FOI.loc[i + pos, binding_site_FOIs] = [\n",
    "            id,\n",
    "            gene_name,\n",
    "            mol_type,\n",
    "            organism,\n",
    "            pre_binding_tmp,\n",
    "            Tm_l,\n",
    "            Tm_r,\n",
    "            True,\n",
    "        ]\n",
    "\n",
    "    pos += pre_binding_num_tmp\n",
    "\n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(file_out_dir + total_pre_binding_file_name, \"a\") as handle:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "    with open(file_out_dir + cut_name, \"a\") as handle:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record[pot_st:pot_en], handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[f\"{id}_{gene_name}\"] = pre_binding_num_tmp\n",
    "\n",
    "with open(tmp + pre_binding_num_file, \"w\") as f:\n",
    "    json.dump(pre_binding_num, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blast and extract blast results\n",
    "\n",
    "NCBIXML: https://homolog.us/Biopython/Bio.Blast.NCBIXML.html#read/0\n",
    "\n",
    "BlastRecord: https://biopython.org/docs/1.75/api/Bio.Blast.Record.html\n",
    "\n",
    "XMLReader: https://codebeautify.org/xmlviewer#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(file_out_dir + total_pre_binding_file_name, \"r\") as f:\n",
    "    fasta_string = f.read()\n",
    "txid = [2697049]  # organism\n",
    "# Submit BLAST search and get handle object\n",
    "handle = NCBIWWW.qblast(\n",
    "    program=\"blastn\",\n",
    "    megablast=\"yes\",\n",
    "    database=\"refseq_rna\",\n",
    "    sequence=fasta_string,\n",
    "    url_base=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi\",\n",
    "    format_object=\"Alignment\",\n",
    "    format_type=\"Xml\",\n",
    ")\n",
    "# read handle object and save to a file\n",
    "with open(tmp + blast_results_file, \"w\") as f:\n",
    "    f.write(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract interested information from blast_results\n",
    "align_num = []\n",
    "\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(tmp + blast_results_file, \"r\") as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split(\"|\")\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        FOI.loc[loca, \"align_accession\"] = \"|\".join(str(_) for _ in align_accession)\n",
    "\n",
    "        # add align_descrip to df\n",
    "        FOI.loc[loca, \"align_descrip\"] = \"|\".join(str(_) for _ in align_descrip_list)\n",
    "\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "\n",
    "        # add plus/minus to df\n",
    "        try:\n",
    "            FOI.loc[loca, \"plus/minus\"] = \",\".join([str(_) for _ in p_m])\n",
    "        except:\n",
    "            FOI.loc[loca, \"plus/minus\"] = \"NAN\"\n",
    "\n",
    "        loca += 1\n",
    "\n",
    "FOI[\"align_num\"][: len(align_num)] = align_num"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select wanted binding site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FOI[\"wanted\"] = [True] * len(FOI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sieve for the suitable binding site\n",
    "gene_name_list = [_.upper() for _ in gene_name_list]\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(FOI)):\n",
    "    # check gene_name\n",
    "    gene_name = FOI.loc[i, \"gene_name\"]\n",
    "    if gene_name.upper() not in gene_name_list:\n",
    "        FOI.loc[i, \"wanted\"] = False\n",
    "    else:\n",
    "        try:\n",
    "            gene_name_list_out.remove(gene_name)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        if FOI.loc[i, \"mol_type\"] != \"mRNA\":\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "    # check gene_organism name\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        spe_ori, gene_ori = FOI.loc[i, \"organism\"], FOI.loc[i, \"gene_name\"]\n",
    "        descrip = FOI.loc[i, \"align_descrip\"].split(\"|\")\n",
    "        for des in descrip:\n",
    "            if gene_ori not in des and spe_ori in des:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "                break\n",
    "\n",
    "    # check plus/minus\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        if pd.isnull(FOI.loc[i, \"plus/minus\"]):\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "        else:\n",
    "            pm_list = FOI.loc[i, \"plus/minus\"].split(\",\")\n",
    "            if \"-1\" not in pm_list:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "    # check G 40%-70%, non consective 5 base\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        seq = FOI.loc[i, \"binding\"]\n",
    "        if \"GGGGG\" in seq:\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "        else:\n",
    "            G_per = seq.count(\"G\") / len(seq)\n",
    "            if G_per < 0.3 or G_per > 0.7:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "    # check Tm\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        Tm_l = FOI.loc[i, \"Tm_l\"]\n",
    "        Tm_r = FOI.loc[i, \"Tm_r\"]\n",
    "        if Tm_l > 65 or Tm_l < 50 or Tm_r > 65 or Tm_r < 50:\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "# write the gene name to search next round\n",
    "with open(output + gene_name_list_tosearch, \"w\") as f:\n",
    "    f.write(\"\\n\".join(gene_name_list_out))\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "FOI.to_excel(tmp + \"probes_sieve.xlsx\")\n",
    "\n",
    "# get the sub dataframe of wanted probes\n",
    "output_df = FOI[FOI[\"wanted\"] == True]\n",
    "# write the output to a xlsx file\n",
    "output_df.to_excel(output + \"probes_wanted.xlsx\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
